// @getbrevo/brevo/dist/esm/api/resources/balance/client/requests/UpdateBalanceDefinitionRequest.mjs
// This file was auto-generated by Fern from our API Definition.
export var UpdateBalanceDefinitionRequest;
(function (UpdateBalanceDefinitionRequest) {
    /** Defines when the balance expires within the selected duration. */
    UpdateBalanceDefinitionRequest.BalanceAvailabilityDurationModifier = {
        NoModification: "noModification",
        StartOfPeriod: "startOfPeriod",
        EndOfPeriod: "endOfPeriod",
    };
    /** Unit of time for balance validity. */
    UpdateBalanceDefinitionRequest.BalanceAvailabilityDurationUnit = {
        Day: "day",
        Week: "week",
        Month: "month",
        Year: "year",
    };
    /** Defines whether partial credit is allowed when reaching max balance. */
    UpdateBalanceDefinitionRequest.BalanceOptionAmountOvertakingStrategy = {
        Strict: "strict",
        Partial: "partial",
    };
    /** Rounding strategy for credit transactions. */
    UpdateBalanceDefinitionRequest.BalanceOptionCreditRounding = {
        Lower: "lower",
        Upper: "upper",
        Natural: "natural",
    };
    /** Rounding strategy for debit transactions. */
    UpdateBalanceDefinitionRequest.BalanceOptionDebitRounding = {
        Lower: "lower",
        Upper: "upper",
        Natural: "natural",
    };
    /** Unit of balance measurement. */
    UpdateBalanceDefinitionRequest.Unit = {
        Points: "POINTS",
        Eur: "EUR",
        Usd: "USD",
        Mxn: "MXN",
        Gbp: "GBP",
        Inr: "INR",
        Cad: "CAD",
        Sgd: "SGD",
        Ron: "RON",
        Jpy: "JPY",
        Myr: "MYR",
        Clp: "CLP",
        Pen: "PEN",
        Mad: "MAD",
        Aud: "AUD",
        Chf: "CHF",
        Brl: "BRL",
    };
})(UpdateBalanceDefinitionRequest || (UpdateBalanceDefinitionRequest = {}));
