// @getbrevo/brevo/dist/esm/api/resources/masterAccount/client/requests/InviteAdminUserRequest.mjs
// This file was auto-generated by Fern from our API Definition.
export var InviteAdminUserRequest;
(function (InviteAdminUserRequest) {
    let Privileges;
    (function (Privileges) {
        let Item;
        (function (Item_1) {
            /** Feature name */
            Item_1.Feature = {
                MyPlan: "my_plan",
                Api: "api",
                UserManagement: "user_management",
                AppManagement: "app_management",
                SubOrganizationGroups: "sub_organization_groups",
                CreateSubOrganizations: "create_sub_organizations",
                ManageSubOrganizations: "manage_sub_organizations",
                Analytics: "analytics",
                Security: "security",
            };
            let Permissions;
            (function (Permissions) {
                Permissions.Item = {
                    All: "all",
                    None: "none",
                    Create: "create",
                    EditDelete: "edit_delete",
                    DownloadData: "download_data",
                    CreateAlerts: "create_alerts",
                };
            })(Permissions = Item_1.Permissions || (Item_1.Permissions = {}));
        })(Item = Privileges.Item || (Privileges.Item = {}));
    })(Privileges = InviteAdminUserRequest.Privileges || (InviteAdminUserRequest.Privileges = {}));
})(InviteAdminUserRequest || (InviteAdminUserRequest = {}));
