// @getbrevo/brevo/dist/esm/api/resources/masterAccount/client/requests/PostCorporateSubAccountSsoTokenRequest.mjs
// This file was auto-generated by Fern from our API Definition.
export var PostCorporateSubAccountSsoTokenRequest;
(function (PostCorporateSubAccountSsoTokenRequest) {
    /**
     * **Set target after login success** * **automation** - Redirect
     * to Automation after login * **email_campaign** - Redirect to
     * Email Campaign after login * **contacts** - Redirect to Contacts
     * after login * **landing_pages** - Redirect to Landing Pages
     * after login * **email_transactional** - Redirect to Email
     * Transactional after login * **senders** - Redirect to Senders
     * after login * **sms_campaign** - Redirect to Sms Campaign after
     * login * **sms_transactional** - Redirect to Sms Transactional
     * after login
     */
    PostCorporateSubAccountSsoTokenRequest.Target = {
        Automation: "automation",
        EmailCampaign: "email_campaign",
        Contacts: "contacts",
        LandingPages: "landing_pages",
        EmailTransactional: "email_transactional",
        Senders: "senders",
        SmsCampaign: "sms_campaign",
        SmsTransactional: "sms_transactional",
    };
})(PostCorporateSubAccountSsoTokenRequest || (PostCorporateSubAccountSsoTokenRequest = {}));
