// @getbrevo/brevo/dist/esm/api/resources/masterAccount/client/requests/PutCorporateUserEmailPermissionsRequest.mjs
// This file was auto-generated by Fern from our API Definition.
export var PutCorporateUserEmailPermissionsRequest;
(function (PutCorporateUserEmailPermissionsRequest) {
    let Privileges;
    (function (Privileges) {
        let Item;
        (function (Item_1) {
            /** feature name */
            Item_1.Feature = {
                UserManagement: "user_management",
                Api: "api",
                MyPlan: "my_plan",
                AppsManagement: "apps_management",
                Analytics: "analytics",
                SubOrganizationGroups: "sub_organization_groups",
                CreateSubOrganizations: "create_sub_organizations",
                ManageSubOrganizations: "manage_sub_organizations",
                Security: "security",
            };
            let Permissions;
            (function (Permissions) {
                Permissions.Item = {
                    All: "all",
                    None: "none",
                    Create: "create",
                    EditDelete: "edit_delete",
                    CreateAlerts: "create_alerts",
                    DownloadData: "download_data",
                    MyLooks: "my_looks",
                    ExploreCreate: "explore_create",
                };
            })(Permissions = Item_1.Permissions || (Item_1.Permissions = {}));
        })(Item = Privileges.Item || (Privileges.Item = {}));
    })(Privileges = PutCorporateUserEmailPermissionsRequest.Privileges || (PutCorporateUserEmailPermissionsRequest.Privileges = {}));
})(PutCorporateUserEmailPermissionsRequest || (PutCorporateUserEmailPermissionsRequest = {}));
