// @getbrevo/brevo/dist/esm/api/types/BalanceDefinition.mjs
// This file was auto-generated by Fern from our API Definition.
export var BalanceDefinition;
(function (BalanceDefinition) {
    /** Unit of balance measurement. */
    BalanceDefinition.Unit = {
        Points: "POINTS",
        Eur: "EUR",
        Usd: "USD",
        Mxn: "MXN",
        Gbp: "GBP",
        Inr: "INR",
        Cad: "CAD",
        Sgd: "SGD",
        Ron: "RON",
        Jpy: "JPY",
        Myr: "MYR",
        Clp: "CLP",
        Pen: "PEN",
        Mad: "MAD",
        Aud: "AUD",
        Chf: "CHF",
        Brl: "BRL",
    };
    /** Partial enables partial credit of balance if maximum balance limit is reached. Strict enables rejection of a transaction if it will breach the max balance limit. */
    BalanceDefinition.BalanceOptionAmountOvertakingStrategy = {
        Strict: "strict",
        Partial: "partial",
    };
    /** Rounding strategy for credit transactions. */
    BalanceDefinition.BalanceOptionCreditRounding = {
        Lower: "lower",
        Upper: "upper",
        Natural: "natural",
    };
    /** Rounding strategy for debit transactions. */
    BalanceDefinition.BalanceOptionDebitRounding = {
        Lower: "lower",
        Upper: "upper",
        Natural: "natural",
    };
    /** Unit of time for the balance availability duration. */
    BalanceDefinition.BalanceAvailabilityDurationUnit = {
        Day: "day",
        Week: "week",
        Month: "month",
        Year: "year",
    };
    /** Modifier for balance availability duration. startOfPeriod causes expiry at the start of the period, endOfPeriod at the end. noModification uses the exact duration from the credit date. */
    BalanceDefinition.BalanceAvailabilityDurationModifier = {
        NoModification: "noModification",
        StartOfPeriod: "startOfPeriod",
        EndOfPeriod: "endOfPeriod",
    };
})(BalanceDefinition || (BalanceDefinition = {}));
