// @getbrevo/brevo/dist/esm/api/types/TierGroup.mjs
// This file was auto-generated by Fern from our API Definition.
export var TierGroup;
(function (TierGroup) {
    /** Select real_time to downgrade tier on real time balance updates. Select membership_anniversary to downgrade tier on subscription anniversary. Select tier_anniversary to downgrade tier on tier anniversary. */
    TierGroup.DowngradeStrategy = {
        RealTime: "real_time",
        MembershipAnniversary: "membership_anniversary",
        TierAnniversary: "tier_anniversary",
    };
    /** Select real_time to upgrade tier on real time balance updates. Select membership_anniversary to upgrade tier on subscription anniversary. Select tier_anniversary to upgrade tier on tier anniversary. */
    TierGroup.UpgradeStrategy = {
        RealTime: "real_time",
        MembershipAnniversary: "membership_anniversary",
        TierAnniversary: "tier_anniversary",
    };
    let UpgradeSchedule;
    (function (UpgradeSchedule) {
        /** Duration unit for the schedule. */
        UpgradeSchedule.DurationUnit = {
            Day: "day",
            Week: "week",
            Month: "month",
            Year: "year",
        };
        /** Modifier for the duration. */
        UpgradeSchedule.DurationModifier = {
            StartOfPeriod: "start_of_period",
            EndOfPeriod: "end_of_period",
        };
    })(UpgradeSchedule = TierGroup.UpgradeSchedule || (TierGroup.UpgradeSchedule = {}));
    let DowngradeSchedule;
    (function (DowngradeSchedule) {
        /** Duration unit for the schedule. */
        DowngradeSchedule.DurationUnit = {
            Day: "day",
            Week: "week",
            Month: "month",
            Year: "year",
        };
        /** Modifier for the duration. */
        DowngradeSchedule.DurationModifier = {
            StartOfPeriod: "start_of_period",
            EndOfPeriod: "end_of_period",
        };
    })(DowngradeSchedule = TierGroup.DowngradeSchedule || (TierGroup.DowngradeSchedule = {}));
})(TierGroup || (TierGroup = {}));
