// @getbrevo/brevo/dist/cjs/api/resources/emailCampaigns/client/requests/CreateEmailCampaignRequest.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateEmailCampaignRequest = void 0;
var CreateEmailCampaignRequest;
(function (CreateEmailCampaignRequest) {
    let EmailExpirationDate;
    (function (EmailExpirationDate) {
        /** unit of the duration */
        EmailExpirationDate.Unit = {
            Days: "days",
            Weeks: "weeks",
            Months: "months",
        };
    })(EmailExpirationDate = CreateEmailCampaignRequest.EmailExpirationDate || (CreateEmailCampaignRequest.EmailExpirationDate = {}));
    /** Choose the metrics that will determinate the winning version. **Mandatory if _splitRule_ >= 1 and < 50**. If splitRule = 50, `winnerCriteria` is ignored if passed */
    CreateEmailCampaignRequest.WinnerCriteria = {
        Open: "open",
        Click: "click",
    };
})(CreateEmailCampaignRequest || (exports.CreateEmailCampaignRequest = CreateEmailCampaignRequest = {}));
