// @getbrevo/brevo/dist/esm/api/resources/tier/client/requests/CreateTierGroupRequest.mjs
// This file was auto-generated by Fern from our API Definition.
export var CreateTierGroupRequest;
(function (CreateTierGroupRequest) {
    /** Select real_time to downgrade tier on real time balance updates. Select membership_anniversary to downgrade tier on subscription anniversary. Select tier_anniversary to downgrade tier on tier anniversary. */
    CreateTierGroupRequest.DowngradeStrategy = {
        RealTime: "real_time",
        MembershipAnniversary: "membership_anniversary",
        TierAnniversary: "tier_anniversary",
    };
    /** Select real_time to upgrade tier on real time balance updates. Select membership_anniversary to upgrade tier on subscription anniversary. Select tier_anniversary to upgrade tier on tier anniversary. */
    CreateTierGroupRequest.UpgradeStrategy = {
        RealTime: "real_time",
        MembershipAnniversary: "membership_anniversary",
        TierAnniversary: "tier_anniversary",
    };
    let UpgradeSchedule;
    (function (UpgradeSchedule) {
        /** Duration unit for the schedule. */
        UpgradeSchedule.DurationUnit = {
            Day: "day",
            Week: "week",
            Month: "month",
            Year: "year",
        };
        /** Modifier for the duration. */
        UpgradeSchedule.DurationModifier = {
            StartOfPeriod: "start_of_period",
            EndOfPeriod: "end_of_period",
        };
    })(UpgradeSchedule = CreateTierGroupRequest.UpgradeSchedule || (CreateTierGroupRequest.UpgradeSchedule = {}));
    let DowngradeSchedule;
    (function (DowngradeSchedule) {
        /** Duration unit for the schedule. */
        DowngradeSchedule.DurationUnit = {
            Day: "day",
            Week: "week",
            Month: "month",
            Year: "year",
        };
        /** Modifier for the duration. */
        DowngradeSchedule.DurationModifier = {
            StartOfPeriod: "start_of_period",
            EndOfPeriod: "end_of_period",
        };
    })(DowngradeSchedule = CreateTierGroupRequest.DowngradeSchedule || (CreateTierGroupRequest.DowngradeSchedule = {}));
})(CreateTierGroupRequest || (CreateTierGroupRequest = {}));
