// @getbrevo/brevo/dist/esm/api/resources/webhooks/client/requests/ExportWebhooksHistoryRequest.mjs
// This file was auto-generated by Fern from our API Definition.
export var ExportWebhooksHistoryRequest;
(function (ExportWebhooksHistoryRequest) {
    /** Filter the history for a specific event type */
    ExportWebhooksHistoryRequest.Event = {
        InvalidParameter: "invalid_parameter",
        MissingParameter: "missing_parameter",
        HardBounce: "hardBounce",
        SoftBounce: "softBounce",
        Delivered: "delivered",
        Spam: "spam",
        Request: "request",
        Opened: "opened",
        Click: "click",
        Invalid: "invalid",
        Deferred: "deferred",
        Blocked: "blocked",
        Unsubscribed: "unsubscribed",
        Error: "error",
        UniqueOpened: "uniqueOpened",
        LoadedByProxy: "loadedByProxy",
        AllEvents: "allEvents",
    };
    /** Filter the history based on webhook type */
    ExportWebhooksHistoryRequest.Type = {
        Transactional: "transactional",
        Marketing: "marketing",
    };
})(ExportWebhooksHistoryRequest || (ExportWebhooksHistoryRequest = {}));
