// @getbrevo/brevo/dist/esm/api/resources/webhooks/client/requests/UpdateWebhookRequest.mjs
// This file was auto-generated by Fern from our API Definition.
export var UpdateWebhookRequest;
(function (UpdateWebhookRequest) {
    let Events;
    (function (Events) {
        Events.Item = {
            Sent: "sent",
            HardBounce: "hardBounce",
            SoftBounce: "softBounce",
            Blocked: "blocked",
            Spam: "spam",
            Delivered: "delivered",
            Request: "request",
            Click: "click",
            Invalid: "invalid",
            Deferred: "deferred",
            Opened: "opened",
            UniqueOpened: "uniqueOpened",
            Unsubscribed: "unsubscribed",
            ListAddition: "listAddition",
            ContactUpdated: "contactUpdated",
            ContactDeleted: "contactDeleted",
            InboundEmailProcessed: "inboundEmailProcessed",
            Reply: "reply",
        };
    })(Events = UpdateWebhookRequest.Events || (UpdateWebhookRequest.Events = {}));
})(UpdateWebhookRequest || (UpdateWebhookRequest = {}));
