// @getbrevo/brevo/dist/cjs/api/resources/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.whatsAppCampaigns = exports.webhooks = exports.wallet = exports.user = exports.transactionalWhatsApp = exports.transactionalSms = exports.transactionalEmails = exports.tier = exports.tasks = exports.smsTemplates = exports.smsCampaigns = exports.senders = exports.reward = exports.program = exports.process = exports.payments = exports.notes = exports.masterAccount = exports.inboundParsing = exports.files = exports.externalFeeds = exports.event = exports.emailCampaigns = exports.ecommerce = exports.domains = exports.deals = exports.customObjects = exports.coupons = exports.conversations = exports.contacts = exports.consentGroups = exports.companies = exports.balance = exports.account = void 0;
__exportStar(require("./account/client/requests/index.js"), exports);
exports.account = __importStar(require("./account/index.js"));
__exportStar(require("./account/types/index.js"), exports);
__exportStar(require("./balance/client/requests/index.js"), exports);
exports.balance = __importStar(require("./balance/index.js"));
__exportStar(require("./balance/types/index.js"), exports);
__exportStar(require("./companies/client/requests/index.js"), exports);
exports.companies = __importStar(require("./companies/index.js"));
__exportStar(require("./companies/types/index.js"), exports);
__exportStar(require("./consentGroups/client/requests/index.js"), exports);
exports.consentGroups = __importStar(require("./consentGroups/index.js"));
__exportStar(require("./consentGroups/types/index.js"), exports);
__exportStar(require("./contacts/client/requests/index.js"), exports);
exports.contacts = __importStar(require("./contacts/index.js"));
__exportStar(require("./contacts/types/index.js"), exports);
__exportStar(require("./conversations/client/requests/index.js"), exports);
exports.conversations = __importStar(require("./conversations/index.js"));
__exportStar(require("./conversations/types/index.js"), exports);
__exportStar(require("./coupons/client/requests/index.js"), exports);
exports.coupons = __importStar(require("./coupons/index.js"));
__exportStar(require("./coupons/types/index.js"), exports);
__exportStar(require("./customObjects/client/requests/index.js"), exports);
exports.customObjects = __importStar(require("./customObjects/index.js"));
__exportStar(require("./customObjects/types/index.js"), exports);
__exportStar(require("./deals/client/requests/index.js"), exports);
exports.deals = __importStar(require("./deals/index.js"));
__exportStar(require("./deals/types/index.js"), exports);
__exportStar(require("./domains/client/requests/index.js"), exports);
exports.domains = __importStar(require("./domains/index.js"));
__exportStar(require("./domains/types/index.js"), exports);
__exportStar(require("./ecommerce/client/requests/index.js"), exports);
exports.ecommerce = __importStar(require("./ecommerce/index.js"));
__exportStar(require("./ecommerce/types/index.js"), exports);
__exportStar(require("./emailCampaigns/client/requests/index.js"), exports);
exports.emailCampaigns = __importStar(require("./emailCampaigns/index.js"));
__exportStar(require("./emailCampaigns/types/index.js"), exports);
__exportStar(require("./event/client/requests/index.js"), exports);
exports.event = __importStar(require("./event/index.js"));
__exportStar(require("./externalFeeds/client/requests/index.js"), exports);
exports.externalFeeds = __importStar(require("./externalFeeds/index.js"));
__exportStar(require("./externalFeeds/types/index.js"), exports);
__exportStar(require("./files/client/requests/index.js"), exports);
exports.files = __importStar(require("./files/index.js"));
__exportStar(require("./files/types/index.js"), exports);
__exportStar(require("./inboundParsing/client/requests/index.js"), exports);
exports.inboundParsing = __importStar(require("./inboundParsing/index.js"));
__exportStar(require("./inboundParsing/types/index.js"), exports);
__exportStar(require("./masterAccount/client/requests/index.js"), exports);
exports.masterAccount = __importStar(require("./masterAccount/index.js"));
__exportStar(require("./masterAccount/types/index.js"), exports);
__exportStar(require("./notes/client/requests/index.js"), exports);
exports.notes = __importStar(require("./notes/index.js"));
__exportStar(require("./notes/types/index.js"), exports);
__exportStar(require("./payments/client/requests/index.js"), exports);
exports.payments = __importStar(require("./payments/index.js"));
__exportStar(require("./payments/types/index.js"), exports);
__exportStar(require("./process/client/requests/index.js"), exports);
exports.process = __importStar(require("./process/index.js"));
__exportStar(require("./process/types/index.js"), exports);
__exportStar(require("./program/client/requests/index.js"), exports);
exports.program = __importStar(require("./program/index.js"));
__exportStar(require("./program/types/index.js"), exports);
__exportStar(require("./reward/client/requests/index.js"), exports);
exports.reward = __importStar(require("./reward/index.js"));
__exportStar(require("./reward/types/index.js"), exports);
__exportStar(require("./senders/client/requests/index.js"), exports);
exports.senders = __importStar(require("./senders/index.js"));
__exportStar(require("./senders/types/index.js"), exports);
__exportStar(require("./smsCampaigns/client/requests/index.js"), exports);
exports.smsCampaigns = __importStar(require("./smsCampaigns/index.js"));
__exportStar(require("./smsCampaigns/types/index.js"), exports);
__exportStar(require("./smsTemplates/client/requests/index.js"), exports);
exports.smsTemplates = __importStar(require("./smsTemplates/index.js"));
__exportStar(require("./smsTemplates/types/index.js"), exports);
__exportStar(require("./tasks/client/requests/index.js"), exports);
exports.tasks = __importStar(require("./tasks/index.js"));
__exportStar(require("./tasks/types/index.js"), exports);
__exportStar(require("./tier/client/requests/index.js"), exports);
exports.tier = __importStar(require("./tier/index.js"));
__exportStar(require("./tier/types/index.js"), exports);
__exportStar(require("./transactionalEmails/client/requests/index.js"), exports);
exports.transactionalEmails = __importStar(require("./transactionalEmails/index.js"));
__exportStar(require("./transactionalEmails/types/index.js"), exports);
__exportStar(require("./transactionalSms/client/requests/index.js"), exports);
exports.transactionalSms = __importStar(require("./transactionalSms/index.js"));
__exportStar(require("./transactionalSms/types/index.js"), exports);
__exportStar(require("./transactionalWhatsApp/client/requests/index.js"), exports);
exports.transactionalWhatsApp = __importStar(require("./transactionalWhatsApp/index.js"));
__exportStar(require("./transactionalWhatsApp/types/index.js"), exports);
__exportStar(require("./user/client/requests/index.js"), exports);
exports.user = __importStar(require("./user/index.js"));
__exportStar(require("./user/types/index.js"), exports);
__exportStar(require("./wallet/client/requests/index.js"), exports);
exports.wallet = __importStar(require("./wallet/index.js"));
__exportStar(require("./webhooks/client/requests/index.js"), exports);
exports.webhooks = __importStar(require("./webhooks/index.js"));
__exportStar(require("./webhooks/types/index.js"), exports);
__exportStar(require("./whatsAppCampaigns/client/requests/index.js"), exports);
exports.whatsAppCampaigns = __importStar(require("./whatsAppCampaigns/index.js"));
__exportStar(require("./whatsAppCampaigns/types/index.js"), exports);
