// @getbrevo/brevo/dist/cjs/api/resources/contacts/client/requests/RequestContactExportRequest.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
Object.defineProperty(exports, "__esModule", { value: true });
exports.RequestContactExportRequest = void 0;
var RequestContactExportRequest;
(function (RequestContactExportRequest) {
    let CustomContactFilter;
    (function (CustomContactFilter) {
        /** **Mandatory if neither actionForEmailCampaigns nor actionForSmsCampaigns is passed.** This will export the contacts on the basis of provided action applied on contacts as per the list id. * **allContacts** - Fetch the list of all contacts for a particular list. * **subscribed & unsubscribed** - Fetch the list of subscribed / unsubscribed (blacklisted via any means) contacts for a particular list. * **unsubscribedPerList** - Fetch the list of contacts that are unsubscribed from a particular list only. */
        CustomContactFilter.ActionForContacts = {
            AllContacts: "allContacts",
            Subscribed: "subscribed",
            Unsubscribed: "unsubscribed",
            UnsubscribedPerList: "unsubscribedPerList",
        };
        /** **Mandatory if neither actionForContacts nor actionForSmsCampaigns is passed.** This will export the contacts on the basis of provided action applied on email campaigns. * **openers & nonOpeners** - emailCampaignId is mandatory. Fetch the list of readers / non-readers for a particular email campaign. * **clickers & nonClickers** - emailCampaignId is mandatory. Fetch the list of clickers / non-clickers for a particular email campaign. * **unsubscribed** - emailCampaignId is mandatory. Fetch the list of all unsubscribed (blacklisted via any means) contacts for a particular email campaign. * **hardBounces & softBounces** - emailCampaignId is optional. Fetch the list of hard bounces / soft bounces for a particular / all email campaign(s). */
        CustomContactFilter.ActionForEmailCampaigns = {
            Openers: "openers",
            NonOpeners: "nonOpeners",
            Clickers: "clickers",
            NonClickers: "nonClickers",
            Unsubscribed: "unsubscribed",
            HardBounces: "hardBounces",
            SoftBounces: "softBounces",
        };
        /** **Mandatory if neither actionForContacts nor actionForEmailCampaigns is passed.** This will export the contacts on the basis of provided action applied on sms campaigns. * **unsubscribed** - Fetch the list of all unsubscribed (blacklisted via any means) contacts for all / particular sms campaigns. * **hardBounces & softBounces** - Fetch the list of hard bounces / soft bounces for all / particular sms campaigns. */
        CustomContactFilter.ActionForSmsCampaigns = {
            HardBounces: "hardBounces",
            SoftBounces: "softBounces",
            Unsubscribed: "unsubscribed",
        };
    })(CustomContactFilter = RequestContactExportRequest.CustomContactFilter || (RequestContactExportRequest.CustomContactFilter = {}));
})(RequestContactExportRequest || (exports.RequestContactExportRequest = RequestContactExportRequest = {}));
