// @getbrevo/brevo/dist/cjs/api/resources/transactionalEmails/types/GetTransacBlockedContactsResponse.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetTransacBlockedContactsResponse = void 0;
var GetTransacBlockedContactsResponse;
(function (GetTransacBlockedContactsResponse) {
    let Contacts;
    (function (Contacts) {
        let Item;
        (function (Item) {
            let Reason;
            (function (Reason) {
                /** Reason code for blocking / unsubscribing (This code is safe for comparison) */
                Reason.Code = {
                    UnsubscribedViaMa: "unsubscribedViaMA",
                    UnsubscribedViaEmail: "unsubscribedViaEmail",
                    AdminBlocked: "adminBlocked",
                    UnsubscribedViaApi: "unsubscribedViaApi",
                    HardBounce: "hardBounce",
                    ContactFlaggedAsSpam: "contactFlaggedAsSpam",
                };
            })(Reason = Item.Reason || (Item.Reason = {}));
        })(Item = Contacts.Item || (Contacts.Item = {}));
    })(Contacts = GetTransacBlockedContactsResponse.Contacts || (GetTransacBlockedContactsResponse.Contacts = {}));
})(GetTransacBlockedContactsResponse || (exports.GetTransacBlockedContactsResponse = GetTransacBlockedContactsResponse = {}));
