// @getbrevo/brevo/dist/cjs/api/resources/transactionalSms/types/GetSmsEventsResponse.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetSmsEventsResponse = void 0;
var GetSmsEventsResponse;
(function (GetSmsEventsResponse) {
    let Events;
    (function (Events) {
        let Item;
        (function (Item) {
            /** Event which occurred */
            Item.Event = {
                Bounces: "bounces",
                HardBounces: "hardBounces",
                SoftBounces: "softBounces",
                Delivered: "delivered",
                Sent: "sent",
                Accepted: "accepted",
                Unsubscription: "unsubscription",
                Replies: "replies",
                Blocked: "blocked",
                Rejected: "rejected",
                Skipped: "skipped",
            };
        })(Item = Events.Item || (Events.Item = {}));
    })(Events = GetSmsEventsResponse.Events || (GetSmsEventsResponse.Events = {}));
})(GetSmsEventsResponse || (exports.GetSmsEventsResponse = GetSmsEventsResponse = {}));
