// @getbrevo/brevo/dist/esm/api/resources/transactionalEmails/types/GetEmailEventReportResponse.mjs
// This file was auto-generated by Fern from our API Definition.
export var GetEmailEventReportResponse;
(function (GetEmailEventReportResponse) {
    let Events;
    (function (Events) {
        let Item;
        (function (Item) {
            /** Event which occurred */
            Item.Event = {
                Bounces: "bounces",
                HardBounces: "hardBounces",
                SoftBounces: "softBounces",
                Delivered: "delivered",
                Spam: "spam",
                Requests: "requests",
                Opened: "opened",
                Clicks: "clicks",
                Invalid: "invalid",
                Deferred: "deferred",
                Blocked: "blocked",
                Unsubscribed: "unsubscribed",
                Error: "error",
                LoadedByProxy: "loadedByProxy",
            };
        })(Item = Events.Item || (Events.Item = {}));
    })(Events = GetEmailEventReportResponse.Events || (GetEmailEventReportResponse.Events = {}));
})(GetEmailEventReportResponse || (GetEmailEventReportResponse = {}));
