// @getbrevo/brevo/dist/cjs/api/resources/account/types/GetAccountResponse.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetAccountResponse = void 0;
var GetAccountResponse;
(function (GetAccountResponse) {
    let Plan;
    (function (Plan) {
        let Item;
        (function (Item) {
            /** Displays the plan type of the user */
            Item.Type = {
                PayAsYouGo: "payAsYouGo",
                Free: "free",
                Subscription: "subscription",
                Sms: "sms",
            };
        })(Item = Plan.Item || (Plan.Item = {}));
    })(Plan = GetAccountResponse.Plan || (GetAccountResponse.Plan = {}));
    let PlanVerticals;
    (function (PlanVerticals) {
        let Item;
        (function (Item) {
            /** Category of the plan */
            Item.PlanCategory = {
                Marketing: "Marketing",
                Chat: "Chat",
                Crm: "CRM",
            };
            /** Current status of the plan */
            Item.Status = {
                Active: "active",
                Inactive: "inactive",
            };
        })(Item = PlanVerticals.Item || (PlanVerticals.Item = {}));
    })(PlanVerticals = GetAccountResponse.PlanVerticals || (GetAccountResponse.PlanVerticals = {}));
})(GetAccountResponse || (exports.GetAccountResponse = GetAccountResponse = {}));
