// @getbrevo/brevo/dist/cjs/api/resources/balance/client/requests/PostLoyaltyBalanceProgramsPidBalanceDefinitionsRequest.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
Object.defineProperty(exports, "__esModule", { value: true });
exports.PostLoyaltyBalanceProgramsPidBalanceDefinitionsRequest = void 0;
var PostLoyaltyBalanceProgramsPidBalanceDefinitionsRequest;
(function (PostLoyaltyBalanceProgramsPidBalanceDefinitionsRequest) {
    /** Defines when the balance expires within the selected duration. */
    PostLoyaltyBalanceProgramsPidBalanceDefinitionsRequest.BalanceAvailabilityDurationModifier = {
        NoModification: "noModification",
        StartOfPeriod: "startOfPeriod",
        EndOfPeriod: "endOfPeriod",
    };
    /** Unit of time for balance validity. */
    PostLoyaltyBalanceProgramsPidBalanceDefinitionsRequest.BalanceAvailabilityDurationUnit = {
        Day: "day",
        Week: "week",
        Month: "month",
        Year: "year",
    };
    /** Defines whether partial credit is allowed when reaching max balance. */
    PostLoyaltyBalanceProgramsPidBalanceDefinitionsRequest.BalanceOptionAmountOvertakingStrategy = {
        Strict: "strict",
        Partial: "partial",
    };
    /** Defines rounding strategy for credit transactions. */
    PostLoyaltyBalanceProgramsPidBalanceDefinitionsRequest.BalanceOptionCreditRounding = {
        Lower: "lower",
        Upper: "upper",
        Natural: "natural",
    };
    /** Defines rounding strategy for debit transactions. */
    PostLoyaltyBalanceProgramsPidBalanceDefinitionsRequest.BalanceOptionDebitRounding = {
        Lower: "lower",
        Upper: "upper",
        Natural: "natural",
    };
    /** Unit of balance measurement. */
    PostLoyaltyBalanceProgramsPidBalanceDefinitionsRequest.Unit = {
        Points: "POINTS",
        Eur: "EUR",
        Usd: "USD",
        Mxn: "MXN",
        Gbp: "GBP",
        Inr: "INR",
        Cad: "CAD",
        Sgd: "SGD",
        Ron: "RON",
        Jpy: "JPY",
        Myr: "MYR",
        Clp: "CLP",
        Pen: "PEN",
        Mad: "MAD",
        Aud: "AUD",
        Chf: "CHF",
        Brl: "BRL",
    };
})(PostLoyaltyBalanceProgramsPidBalanceDefinitionsRequest || (exports.PostLoyaltyBalanceProgramsPidBalanceDefinitionsRequest = PostLoyaltyBalanceProgramsPidBalanceDefinitionsRequest = {}));
