// @getbrevo/brevo/dist/cjs/api/resources/customObjects/client/Client.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomObjectsClient = void 0;
const BaseClient_js_1 = require("../../../../BaseClient.js");
const headers_js_1 = require("../../../../core/headers.js");
const core = __importStar(require("../../../../core/index.js"));
const environments = __importStar(require("../../../../environments.js"));
const handleNonStatusCodeError_js_1 = require("../../../../errors/handleNonStatusCodeError.js");
const errors = __importStar(require("../../../../errors/index.js"));
const Brevo = __importStar(require("../../../index.js"));
class CustomObjectsClient {
    constructor(options) {
        this._options = (0, BaseClient_js_1.normalizeClientOptionsWithAuth)(options);
    }
    /**
     * <Note title="Enterprise access only">Custom objects are only available to Enterprise plans.
     * This feature is in beta. These are subject to change.</Note>
     * Performs bulk create or update (upsert) operations for object records in a single asynchronous request. This endpoint is optimized for high-volume data imports and synchronization scenarios.
     *
     * **How Upsert Works:**
     * - **Create**: Omit `identifiers`, or provide only `ext_id` (if it doesn't already exist). A new record is created with a Brevo-generated `id`.
     * - **Update**: Provide `id` (Brevo internal ID) or an `ext_id` that already exists. The matching record is updated with the new attribute values.
     * - **Important:** `id` is for **updates only**. Providing an `id` that does not belong to an existing record will fail during async processing (the HTTP response will still be 202, but the record will be rejected in the background). To create a new record with a stable external reference, use `ext_id` instead.
     *
     * **Request Structure:**
     * Each object record in the `records` array can include:
     * - `identifiers`: Either `id` (internal Brevo ID) or `ext_id` (your external system ID) — required for updates. **Note:** use `id` (singular), not `ids`.
     * - `attributes`: Key-value pairs where each key is the attribute **key** (e.g., `company_name`), not the attribute label (e.g., "Company Name").
     * - `associations`: Controls linking and unlinking of associated records (optional). Each entry specifies:
     *     - `object_type`: The type of the associated object
     *     - `action`: `link` (default) to create the association, or `unlink` to remove it
     *     - `records`: The associated records to link or unlink (each identified by `ext_id` or `id`)
     *     - **Unlink is idempotent** — unlinking a non-existing association is a no-op (no error returned)
     *     - `link` and `unlink` actions can be submitted for the same `object_type` in a single record entry
     *     - Both associated records must already exist before a link can be created
     *
     * > **Common mistake:** Passing the attribute **label** (the display name you see in the UI) instead of the attribute **key** will cause the attribute to be silently ignored and the record may not be created as expected.
     *
     * **Asynchronous Processing:**
     * - Returns immediately with a `processId` (HTTP 202 Accepted)
     * - Use the processId to track status via the Get process API
     *
     * **API and Schema Limitations:**
     * - Max 1000 object records per request
     * - Max request body size: 1 MB
     * - Max 500 attributes per object record (matches the schema limit of 500 attributes per object)
     * - Unknown attribute keys are silently ignored (no error, no attribute creation)
     * - Max 10 association records per associated object-type in each record of the request. If you need more, send multiple requests.
     *
     * **Important Behaviors:**
     * - The object schema must be created before upserting records
     * - Unknown attribute keys are silently ignored (no error, no creation)
     * - Both associated object records must already exist before creating a link association
     * - Unlink operations are idempotent: attempting to unlink a non-existing association returns success
     * - `link` and `unlink` actions can be submitted for the same `object_type` in a single record entry
     * - Contact objects cannot be created via this endpoint
     * - For `category` and `multiple_category` attributes, pass the option **key** as the value (not the option label or option ID).
     * - The `id` identifier (internal Brevo ID) can only be used for **updating** existing records. To create new records, either omit identifiers (Brevo auto-generates an ID) or provide an `ext_id`.
     *
     * **Errors:**
     * - Make sure both object records exist before associating them, else the API will return an error.
     * - This route does not create objects. The object where the object records are upserted by this API must be created already else the API will return an error "invalid object type".
     *
     * @param {Brevo.UpsertrecordsRequest} request
     * @param {CustomObjectsClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link Brevo.BadRequestError}
     * @throws {@link Brevo.ForbiddenError}
     * @throws {@link Brevo.NotFoundError}
     * @throws {@link Brevo.InternalServerError}
     *
     * @example
     *     await client.customObjects.upsertrecords({
     *         object_type: "vehicle",
     *         records: [{
     *                 associations: [{
     *                         object_type: "garage",
     *                         action: "link",
     *                         records: [{
     *                                 identifiers: {
     *                                     id: 435435
     *                                 }
     *                             }]
     *                     }, {
     *                         object_type: "garage",
     *                         action: "unlink",
     *                         records: [{
     *                                 identifiers: {
     *                                     ext_id: "old-garage-001"
     *                                 }
     *                             }]
     *                     }],
     *                 attributes: {
     *                     "make": "Toyota",
     *                     "model": "Camry",
     *                     "year": 2020,
     *                     "engine_type": "hybrid"
     *                 },
     *                 identifiers: {
     *                     ext_id: "VIN123"
     *                 }
     *             }]
     *     })
     *
     * @example
     *     await client.customObjects.upsertrecords({
     *         object_type: "vehicle",
     *         records: [{
     *                 attributes: {
     *                     "make": "Honda",
     *                     "model": "Civic",
     *                     "year": 2023,
     *                     "engine_type": "petrol"
     *                 }
     *             }]
     *     })
     *
     * @example
     *     await client.customObjects.upsertrecords({
     *         object_type: "vehicle",
     *         records: [{
     *                 attributes: {
     *                     "color": "red",
     *                     "engine_type": "diesel"
     *                 },
     *                 identifiers: {
     *                     id: 42
     *                 }
     *             }]
     *     })
     */
    upsertrecords(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__upsertrecords(request, requestOptions));
    }
    __upsertrecords(request, requestOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b, _c, _d, _e, _f, _g, _h, _j;
            const { object_type: objectType } = request, _body = __rest(request, ["object_type"]);
            const _authRequest = yield this._options.authProvider.getAuthRequest();
            const _headers = (0, headers_js_1.mergeHeaders)(_authRequest.headers, (_a = this._options) === null || _a === void 0 ? void 0 : _a.headers, requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.headers);
            const _response = yield core.fetcher({
                url: core.url.join((_c = (_b = (yield core.Supplier.get(this._options.baseUrl))) !== null && _b !== void 0 ? _b : (yield core.Supplier.get(this._options.environment))) !== null && _c !== void 0 ? _c : environments.BrevoEnvironment.Default, `objects/${core.url.encodePathParam(objectType)}/batch/upsert`),
                method: "POST",
                headers: _headers,
                contentType: "application/json",
                queryString: core.url.queryBuilder().mergeAdditional(requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.queryParams).build(),
                requestType: "json",
                body: _body,
                timeoutMs: ((_f = (_d = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.timeoutInSeconds) !== null && _d !== void 0 ? _d : (_e = this._options) === null || _e === void 0 ? void 0 : _e.timeoutInSeconds) !== null && _f !== void 0 ? _f : 60) * 1000,
                maxRetries: (_g = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.maxRetries) !== null && _g !== void 0 ? _g : (_h = this._options) === null || _h === void 0 ? void 0 : _h.maxRetries,
                abortSignal: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.abortSignal,
                fetchFn: (_j = this._options) === null || _j === void 0 ? void 0 : _j.fetch,
                logging: this._options.logging,
            });
            if (_response.ok) {
                return { data: _response.body, rawResponse: _response.rawResponse };
            }
            if (_response.error.reason === "status-code") {
                switch (_response.error.statusCode) {
                    case 400:
                        throw new Brevo.BadRequestError(_response.error.body, _response.rawResponse);
                    case 403:
                        throw new Brevo.ForbiddenError(_response.error.body, _response.rawResponse);
                    case 404:
                        throw new Brevo.NotFoundError(_response.error.body, _response.rawResponse);
                    case 500:
                        throw new Brevo.InternalServerError(_response.error.body, _response.rawResponse);
                    default:
                        throw new errors.BrevoError({
                            statusCode: _response.error.statusCode,
                            body: _response.error.body,
                            rawResponse: _response.rawResponse,
                        });
                }
            }
            return (0, handleNonStatusCodeError_js_1.handleNonStatusCodeError)(_response.error, _response.rawResponse, "POST", "/objects/{object_type}/batch/upsert");
        });
    }
    /**
     * <Note title="Enterprise access only">Custom objects are only available to Enterprise plans.
     * This feature is in beta. These are subject to change.</Note>
     * This API retrieves a list of object records along with their associated records and provides the total count of records for the specified object. **Note**: Contact as object type is not supported in this endpoint.
     *
     * @param {Brevo.GetrecordsRequest} request
     * @param {CustomObjectsClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link Brevo.BadRequestError}
     * @throws {@link Brevo.ForbiddenError}
     * @throws {@link Brevo.FailedDependencyError}
     * @throws {@link Brevo.InternalServerError}
     *
     * @example
     *     await client.customObjects.getrecords({
     *         object_type: "vehicle",
     *         limit: 1000000,
     *         page_num: 1000000
     *     })
     */
    getrecords(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__getrecords(request, requestOptions));
    }
    __getrecords(request, requestOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b, _c, _d, _e, _f, _g, _h, _j;
            const { object_type: objectType, limit, page_num: pageNum, sort, association } = request;
            const _queryParams = {
                limit,
                page_num: pageNum,
                sort: sort != null ? sort : undefined,
                association: association != null ? association : undefined,
            };
            const _authRequest = yield this._options.authProvider.getAuthRequest();
            const _headers = (0, headers_js_1.mergeHeaders)(_authRequest.headers, (_a = this._options) === null || _a === void 0 ? void 0 : _a.headers, requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.headers);
            const _response = yield core.fetcher({
                url: core.url.join((_c = (_b = (yield core.Supplier.get(this._options.baseUrl))) !== null && _b !== void 0 ? _b : (yield core.Supplier.get(this._options.environment))) !== null && _c !== void 0 ? _c : environments.BrevoEnvironment.Default, `objects/${core.url.encodePathParam(objectType)}/records`),
                method: "GET",
                headers: _headers,
                queryString: core.url
                    .queryBuilder()
                    .addMany(_queryParams)
                    .mergeAdditional(requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.queryParams)
                    .build(),
                timeoutMs: ((_f = (_d = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.timeoutInSeconds) !== null && _d !== void 0 ? _d : (_e = this._options) === null || _e === void 0 ? void 0 : _e.timeoutInSeconds) !== null && _f !== void 0 ? _f : 60) * 1000,
                maxRetries: (_g = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.maxRetries) !== null && _g !== void 0 ? _g : (_h = this._options) === null || _h === void 0 ? void 0 : _h.maxRetries,
                abortSignal: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.abortSignal,
                fetchFn: (_j = this._options) === null || _j === void 0 ? void 0 : _j.fetch,
                logging: this._options.logging,
            });
            if (_response.ok) {
                return { data: _response.body, rawResponse: _response.rawResponse };
            }
            if (_response.error.reason === "status-code") {
                switch (_response.error.statusCode) {
                    case 400:
                        throw new Brevo.BadRequestError(_response.error.body, _response.rawResponse);
                    case 403:
                        throw new Brevo.ForbiddenError(_response.error.body, _response.rawResponse);
                    case 424:
                        throw new Brevo.FailedDependencyError(_response.error.body, _response.rawResponse);
                    case 500:
                        throw new Brevo.InternalServerError(_response.error.body, _response.rawResponse);
                    default:
                        throw new errors.BrevoError({
                            statusCode: _response.error.statusCode,
                            body: _response.error.body,
                            rawResponse: _response.rawResponse,
                        });
                }
            }
            return (0, handleNonStatusCodeError_js_1.handleNonStatusCodeError)(_response.error, _response.rawResponse, "GET", "/objects/{object_type}/records");
        });
    }
    /**
     * Use this endpoint to delete multiple object records of the same object-type in one request.
     * The request is accepted and processed asynchronously. You can track the status of the deletion process using the returned **processId**.
     * **Limitations:** - Each request can contain up to **1000** object record identifiers - Either `ids` or `ext_ids` must be provided, but **not both** in the same request - Deletion of Brevo standard object records is not supported via this endpoint - If more records must be deleted, send multiple batch requests
     *
     * @param {Brevo.BatchDeleteObjectRecordsRequest} request
     * @param {CustomObjectsClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link Brevo.BadRequestError}
     * @throws {@link Brevo.ForbiddenError}
     * @throws {@link Brevo.InternalServerError}
     *
     * @example
     *     await client.customObjects.batchDeleteObjectRecords({
     *         object_type: "vehicle",
     *         identifiers: {
     *             ext_ids: ["ext-001", "ext-002"]
     *         }
     *     })
     */
    batchDeleteObjectRecords(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__batchDeleteObjectRecords(request, requestOptions));
    }
    __batchDeleteObjectRecords(request, requestOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b, _c, _d, _e, _f, _g, _h, _j;
            const { object_type: objectType } = request, _body = __rest(request, ["object_type"]);
            const _authRequest = yield this._options.authProvider.getAuthRequest();
            const _headers = (0, headers_js_1.mergeHeaders)(_authRequest.headers, (_a = this._options) === null || _a === void 0 ? void 0 : _a.headers, requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.headers);
            const _response = yield core.fetcher({
                url: core.url.join((_c = (_b = (yield core.Supplier.get(this._options.baseUrl))) !== null && _b !== void 0 ? _b : (yield core.Supplier.get(this._options.environment))) !== null && _c !== void 0 ? _c : environments.BrevoEnvironment.Default, `objects/${core.url.encodePathParam(objectType)}/batch/delete`),
                method: "POST",
                headers: _headers,
                contentType: "application/json",
                queryString: core.url.queryBuilder().mergeAdditional(requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.queryParams).build(),
                requestType: "json",
                body: _body,
                timeoutMs: ((_f = (_d = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.timeoutInSeconds) !== null && _d !== void 0 ? _d : (_e = this._options) === null || _e === void 0 ? void 0 : _e.timeoutInSeconds) !== null && _f !== void 0 ? _f : 60) * 1000,
                maxRetries: (_g = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.maxRetries) !== null && _g !== void 0 ? _g : (_h = this._options) === null || _h === void 0 ? void 0 : _h.maxRetries,
                abortSignal: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.abortSignal,
                fetchFn: (_j = this._options) === null || _j === void 0 ? void 0 : _j.fetch,
                logging: this._options.logging,
            });
            if (_response.ok) {
                return {
                    data: _response.body,
                    rawResponse: _response.rawResponse,
                };
            }
            if (_response.error.reason === "status-code") {
                switch (_response.error.statusCode) {
                    case 400:
                        throw new Brevo.BadRequestError(_response.error.body, _response.rawResponse);
                    case 403:
                        throw new Brevo.ForbiddenError(_response.error.body, _response.rawResponse);
                    case 500:
                        throw new Brevo.InternalServerError(_response.error.body, _response.rawResponse);
                    default:
                        throw new errors.BrevoError({
                            statusCode: _response.error.statusCode,
                            body: _response.error.body,
                            rawResponse: _response.rawResponse,
                        });
                }
            }
            return (0, handleNonStatusCodeError_js_1.handleNonStatusCodeError)(_response.error, _response.rawResponse, "POST", "/objects/{object_type}/batch/delete");
        });
    }
}
exports.CustomObjectsClient = CustomObjectsClient;
