// @getbrevo/brevo/dist/esm/errors/handleNonStatusCodeError.mjs
// This file was auto-generated by Fern from our API Definition.
import * as errors from "./index.mjs";
export function handleNonStatusCodeError(error, rawResponse, method, path) {
    switch (error.reason) {
        case "non-json":
            throw new errors.BrevoError({
                statusCode: error.statusCode,
                body: error.rawBody,
                rawResponse: rawResponse,
            });
        case "body-is-null":
            throw new errors.BrevoError({
                statusCode: error.statusCode,
                rawResponse: rawResponse,
            });
        case "timeout":
            throw new errors.BrevoTimeoutError(`Timeout exceeded when calling ${method} ${path}.`, {
                cause: error.cause,
            });
        case "unknown":
            throw new errors.BrevoError({
                message: error.errorMessage,
                rawResponse: rawResponse,
                cause: error.cause,
            });
        default:
            throw new errors.BrevoError({
                message: "Unknown error",
                rawResponse: rawResponse,
            });
    }
}
