// @getbrevo/brevo/dist/esm/api/resources/process/types/GetProcessesResponse.mjs
// This file was auto-generated by Fern from our API Definition.
export var GetProcessesResponse;
(function (GetProcessesResponse) {
    let Processes;
    (function (Processes) {
        let Item;
        (function (Item) {
            /** Name/type of the process */
            Item.Name = {
                Importuser: "IMPORTUSER",
                SearchExportUsers: "SEARCH_EXPORT_USERS",
                TransCalc: "TRANS-CALC",
                TransGlobalCalc: "TRANS-GLOBAL-CALC",
                CampaignProcessing: "CAMPAIGN_PROCESSING",
                ListExport: "LIST_EXPORT",
                ContactExport: "CONTACT_EXPORT",
            };
            /** Current status of the process */
            Item.Status = {
                Queued: "queued",
                InProcess: "in_process",
                Processing: "processing",
                Completed: "completed",
                Failed: "failed",
                Cancelled: "cancelled",
            };
        })(Item = Processes.Item || (Processes.Item = {}));
    })(Processes = GetProcessesResponse.Processes || (GetProcessesResponse.Processes = {}));
})(GetProcessesResponse || (GetProcessesResponse = {}));
