// @getbrevo/brevo/dist/esm/api/resources/externalFeeds/client/Client.mjs
// This file was auto-generated by Fern from our API Definition.
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { normalizeClientOptionsWithAuth } from "../../../../BaseClient.mjs";
import { mergeHeaders } from "../../../../core/headers.mjs";
import * as core from "../../../../core/index.mjs";
import * as environments from "../../../../environments.mjs";
import { handleNonStatusCodeError } from "../../../../errors/handleNonStatusCodeError.mjs";
import * as errors from "../../../../errors/index.mjs";
import * as Brevo from "../../../index.mjs";
export class ExternalFeedsClient {
    constructor(options) {
        this._options = normalizeClientOptionsWithAuth(options);
    }
    /**
     * Retrieves all external feeds from your Brevo account with filtering and pagination.
     *
     * **Use this to:**
     * - Get an overview of all external data feeds
     * - Find feeds by name using search functionality
     * - Filter feeds by creation date range
     * - Browse feeds by authentication type
     * - Monitor feed library organization and usage
     *
     * **Key information returned:**
     * - Feed details (UUID, name, URL, authentication type)
     * - Feed configuration and settings
     * - Creation and modification timestamps
     * - Feed status and error information
     * - Authentication and header configurations
     *
     * **Important considerations:**
     * - External feeds enable dynamic content in email campaigns
     * - Feeds must be accessible from Brevo servers
     * - Authentication credentials are securely stored
     * - Feed performance affects campaign delivery
     * - Use pagination for accounts with many feeds
     * - Date range filtering limited to 30 days maximum
     * - Search functionality works on feed name matching
     * - Internal feeds are system-managed and cannot be modified
     *
     * @param {Brevo.GetAllExternalFeedsRequest} request
     * @param {ExternalFeedsClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link Brevo.BadRequestError}
     * @throws {@link Brevo.NotFoundError}
     *
     * @example
     *     await client.externalFeeds.getAllExternalFeeds({
     *         search: "product",
     *         startDate: "2024-01-01",
     *         endDate: "2024-01-31"
     *     })
     */
    getAllExternalFeeds(request = {}, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__getAllExternalFeeds(request, requestOptions));
    }
    __getAllExternalFeeds() {
        return __awaiter(this, arguments, void 0, function* (request = {}, requestOptions) {
            var _a, _b, _c, _d, _e, _f, _g, _h, _j;
            const { search, startDate, endDate, sort, authType, limit, offset } = request;
            const _queryParams = {
                search,
                startDate: startDate != null ? startDate : undefined,
                endDate: endDate != null ? endDate : undefined,
                sort: sort != null ? sort : undefined,
                authType: authType != null ? authType : undefined,
                limit,
                offset,
            };
            const _authRequest = yield this._options.authProvider.getAuthRequest();
            const _headers = mergeHeaders(_authRequest.headers, (_a = this._options) === null || _a === void 0 ? void 0 : _a.headers, requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.headers);
            const _response = yield core.fetcher({
                url: core.url.join((_c = (_b = (yield core.Supplier.get(this._options.baseUrl))) !== null && _b !== void 0 ? _b : (yield core.Supplier.get(this._options.environment))) !== null && _c !== void 0 ? _c : environments.BrevoEnvironment.Default, "feeds"),
                method: "GET",
                headers: _headers,
                queryString: core.url
                    .queryBuilder()
                    .addMany(_queryParams)
                    .mergeAdditional(requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.queryParams)
                    .build(),
                timeoutMs: ((_f = (_d = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.timeoutInSeconds) !== null && _d !== void 0 ? _d : (_e = this._options) === null || _e === void 0 ? void 0 : _e.timeoutInSeconds) !== null && _f !== void 0 ? _f : 60) * 1000,
                maxRetries: (_g = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.maxRetries) !== null && _g !== void 0 ? _g : (_h = this._options) === null || _h === void 0 ? void 0 : _h.maxRetries,
                abortSignal: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.abortSignal,
                fetchFn: (_j = this._options) === null || _j === void 0 ? void 0 : _j.fetch,
                logging: this._options.logging,
            });
            if (_response.ok) {
                return { data: _response.body, rawResponse: _response.rawResponse };
            }
            if (_response.error.reason === "status-code") {
                switch (_response.error.statusCode) {
                    case 400:
                        throw new Brevo.BadRequestError(_response.error.body, _response.rawResponse);
                    case 404:
                        throw new Brevo.NotFoundError(_response.error.body, _response.rawResponse);
                    default:
                        throw new errors.BrevoError({
                            statusCode: _response.error.statusCode,
                            body: _response.error.body,
                            rawResponse: _response.rawResponse,
                        });
                }
            }
            return handleNonStatusCodeError(_response.error, _response.rawResponse, "GET", "/feeds");
        });
    }
    /**
     * Creates a new external feed for dynamic content in email campaigns.
     *
     * **Use this to:**
     * - Set up external data sources for dynamic content
     * - Configure authentication for protected feeds
     * - Enable real-time content updates in campaigns
     * - Establish connections to product catalogs, blogs, or APIs
     *
     * **Key information returned:**
     * - Created feed UUID for reference in campaigns
     * - Success confirmation
     *
     * **Important considerations:**
     * - Feed URL must be accessible from Brevo infrastructure
     * - Authentication credentials are securely encrypted
     * - Test feed accessibility before campaign use
     * - Consider feed response time for campaign performance
     * - Monitor feed reliability and uptime
     * - Use caching for frequently accessed feeds
     * - Maximum 5 retry attempts allowed for failed requests
     * - Custom headers support for API integration requirements
     *
     * @param {Brevo.CreateExternalFeedRequest} request
     * @param {ExternalFeedsClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link Brevo.BadRequestError}
     *
     * @example
     *     await client.externalFeeds.createExternalFeed({
     *         name: "Public API Feed",
     *         url: "https://jsonplaceholder.typicode.com/posts",
     *         authType: "noAuth",
     *         maxRetries: 3,
     *         cache: true
     *     })
     *
     * @example
     *     await client.externalFeeds.createExternalFeed({
     *         name: "Product Catalog",
     *         url: "https://api.example.com/products",
     *         authType: "noAuth",
     *         maxRetries: 3,
     *         cache: true
     *     })
     *
     * @example
     *     await client.externalFeeds.createExternalFeed({
     *         name: "Internal Blog Feed",
     *         url: "https://blog.example.com/api/posts",
     *         authType: "noAuth",
     *         maxRetries: 3,
     *         cache: true
     *     })
     */
    createExternalFeed(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__createExternalFeed(request, requestOptions));
    }
    __createExternalFeed(request, requestOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b, _c, _d, _e, _f, _g, _h, _j;
            const _authRequest = yield this._options.authProvider.getAuthRequest();
            const _headers = mergeHeaders(_authRequest.headers, (_a = this._options) === null || _a === void 0 ? void 0 : _a.headers, requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.headers);
            const _response = yield core.fetcher({
                url: core.url.join((_c = (_b = (yield core.Supplier.get(this._options.baseUrl))) !== null && _b !== void 0 ? _b : (yield core.Supplier.get(this._options.environment))) !== null && _c !== void 0 ? _c : environments.BrevoEnvironment.Default, "feeds"),
                method: "POST",
                headers: _headers,
                contentType: "application/json",
                queryString: core.url.queryBuilder().mergeAdditional(requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.queryParams).build(),
                requestType: "json",
                body: request,
                timeoutMs: ((_f = (_d = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.timeoutInSeconds) !== null && _d !== void 0 ? _d : (_e = this._options) === null || _e === void 0 ? void 0 : _e.timeoutInSeconds) !== null && _f !== void 0 ? _f : 60) * 1000,
                maxRetries: (_g = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.maxRetries) !== null && _g !== void 0 ? _g : (_h = this._options) === null || _h === void 0 ? void 0 : _h.maxRetries,
                abortSignal: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.abortSignal,
                fetchFn: (_j = this._options) === null || _j === void 0 ? void 0 : _j.fetch,
                logging: this._options.logging,
            });
            if (_response.ok) {
                return { data: _response.body, rawResponse: _response.rawResponse };
            }
            if (_response.error.reason === "status-code") {
                switch (_response.error.statusCode) {
                    case 400:
                        throw new Brevo.BadRequestError(_response.error.body, _response.rawResponse);
                    default:
                        throw new errors.BrevoError({
                            statusCode: _response.error.statusCode,
                            body: _response.error.body,
                            rawResponse: _response.rawResponse,
                        });
                }
            }
            return handleNonStatusCodeError(_response.error, _response.rawResponse, "POST", "/feeds");
        });
    }
    /**
     * Retrieves details of a specific external feed by its UUID.
     *
     * **Use this to:**
     * - Get complete configuration of an external feed
     * - Check feed authentication settings
     * - Review feed personalization options
     * - Verify feed URL and parameters
     * - Monitor feed modification history
     *
     * **Key information returned:**
     * - Complete feed configuration and settings
     * - Authentication credentials and headers
     * - Personalization and fallback settings
     * - Creation and modification timestamps
     * - Cache and retry configurations
     *
     * **Important considerations:**
     * - UUID must exist in your account
     * - Provides complete feed information for troubleshooting
     * - Essential before making modifications
     * - Shows current feed health status
     * - Useful for debugging feed issues
     *
     * @param {Brevo.GetExternalFeedByUuidRequest} request
     * @param {ExternalFeedsClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link Brevo.BadRequestError}
     * @throws {@link Brevo.NotFoundError}
     *
     * @example
     *     await client.externalFeeds.getExternalFeedByUuid({
     *         uuid: "b1c2d3e4-f5a6-47b8-89c0-d1e2f3a4b5c6"
     *     })
     */
    getExternalFeedByUuid(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__getExternalFeedByUuid(request, requestOptions));
    }
    __getExternalFeedByUuid(request, requestOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b, _c, _d, _e, _f, _g, _h, _j;
            const { uuid } = request;
            const _authRequest = yield this._options.authProvider.getAuthRequest();
            const _headers = mergeHeaders(_authRequest.headers, (_a = this._options) === null || _a === void 0 ? void 0 : _a.headers, requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.headers);
            const _response = yield core.fetcher({
                url: core.url.join((_c = (_b = (yield core.Supplier.get(this._options.baseUrl))) !== null && _b !== void 0 ? _b : (yield core.Supplier.get(this._options.environment))) !== null && _c !== void 0 ? _c : environments.BrevoEnvironment.Default, `feeds/${core.url.encodePathParam(uuid)}`),
                method: "GET",
                headers: _headers,
                queryString: core.url.queryBuilder().mergeAdditional(requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.queryParams).build(),
                timeoutMs: ((_f = (_d = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.timeoutInSeconds) !== null && _d !== void 0 ? _d : (_e = this._options) === null || _e === void 0 ? void 0 : _e.timeoutInSeconds) !== null && _f !== void 0 ? _f : 60) * 1000,
                maxRetries: (_g = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.maxRetries) !== null && _g !== void 0 ? _g : (_h = this._options) === null || _h === void 0 ? void 0 : _h.maxRetries,
                abortSignal: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.abortSignal,
                fetchFn: (_j = this._options) === null || _j === void 0 ? void 0 : _j.fetch,
                logging: this._options.logging,
            });
            if (_response.ok) {
                return { data: _response.body, rawResponse: _response.rawResponse };
            }
            if (_response.error.reason === "status-code") {
                switch (_response.error.statusCode) {
                    case 400:
                        throw new Brevo.BadRequestError(_response.error.body, _response.rawResponse);
                    case 404:
                        throw new Brevo.NotFoundError(_response.error.body, _response.rawResponse);
                    default:
                        throw new errors.BrevoError({
                            statusCode: _response.error.statusCode,
                            body: _response.error.body,
                            rawResponse: _response.rawResponse,
                        });
                }
            }
            return handleNonStatusCodeError(_response.error, _response.rawResponse, "GET", "/feeds/{uuid}");
        });
    }
    /**
     * Updates configuration of an existing external feed.
     *
     * **Use this to:**
     * - Update feed URLs when data sources change
     * - Modify authentication credentials
     * - Change cache and retry settings
     * - Update custom headers
     * - Rename feeds for better organization
     *
     * **Key information returned:**
     * - Success confirmation message
     *
     * **Important considerations:**
     * - Only provided fields will be updated
     * - Feed UUID must exist in your account
     * - Authentication changes require verification
     * - URL changes should be tested before campaign use
     * - Monitor campaigns using this feed after updates
     *
     * @param {Brevo.UpdateExternalFeedRequest} request
     * @param {ExternalFeedsClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link Brevo.BadRequestError}
     * @throws {@link Brevo.NotFoundError}
     *
     * @example
     *     await client.externalFeeds.updateExternalFeed({
     *         uuid: "b1c2d3e4-f5a6-47b8-89c0-d1e2f3a4b5c6",
     *         name: "Updated Product Catalog",
     *         url: "https://api.newstore.com/products/v2"
     *     })
     *
     * @example
     *     await client.externalFeeds.updateExternalFeed({
     *         uuid: "b1c2d3e4-f5a6-47b8-89c0-d1e2f3a4b5c6",
     *         authType: "token",
     *         token: "newabc123token789"
     *     })
     *
     * @example
     *     await client.externalFeeds.updateExternalFeed({
     *         uuid: "b1c2d3e4-f5a6-47b8-89c0-d1e2f3a4b5c6",
     *         maxRetries: 5,
     *         cache: true
     *     })
     *
     * @example
     *     await client.externalFeeds.updateExternalFeed({
     *         uuid: "b1c2d3e4-f5a6-47b8-89c0-d1e2f3a4b5c6",
     *         headers: [{
     *                 name: "X-API-Version",
     *                 value: "v2"
     *             }, {
     *                 name: "User-Agent",
     *                 value: "Brevo-Feed-Client/1.0"
     *             }]
     *     })
     *
     * @example
     *     await client.externalFeeds.updateExternalFeed({
     *         uuid: "b1c2d3e4-f5a6-47b8-89c0-d1e2f3a4b5c6",
     *         name: "Enterprise Product Feed",
     *         url: "https://enterprise-api.company.com/catalog",
     *         authType: "token",
     *         token: "enterprise789token123",
     *         headers: [{
     *                 name: "Authorization",
     *                 value: "Bearer enterprise789token123"
     *             }, {
     *                 name: "Accept",
     *                 value: "application/json"
     *             }],
     *         maxRetries: 3,
     *         cache: true
     *     })
     */
    updateExternalFeed(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__updateExternalFeed(request, requestOptions));
    }
    __updateExternalFeed(request, requestOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b, _c, _d, _e, _f, _g, _h, _j;
            const { uuid } = request, _body = __rest(request, ["uuid"]);
            const _authRequest = yield this._options.authProvider.getAuthRequest();
            const _headers = mergeHeaders(_authRequest.headers, (_a = this._options) === null || _a === void 0 ? void 0 : _a.headers, requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.headers);
            const _response = yield core.fetcher({
                url: core.url.join((_c = (_b = (yield core.Supplier.get(this._options.baseUrl))) !== null && _b !== void 0 ? _b : (yield core.Supplier.get(this._options.environment))) !== null && _c !== void 0 ? _c : environments.BrevoEnvironment.Default, `feeds/${core.url.encodePathParam(uuid)}`),
                method: "PUT",
                headers: _headers,
                contentType: "application/json",
                queryString: core.url.queryBuilder().mergeAdditional(requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.queryParams).build(),
                requestType: "json",
                body: _body,
                timeoutMs: ((_f = (_d = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.timeoutInSeconds) !== null && _d !== void 0 ? _d : (_e = this._options) === null || _e === void 0 ? void 0 : _e.timeoutInSeconds) !== null && _f !== void 0 ? _f : 60) * 1000,
                maxRetries: (_g = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.maxRetries) !== null && _g !== void 0 ? _g : (_h = this._options) === null || _h === void 0 ? void 0 : _h.maxRetries,
                abortSignal: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.abortSignal,
                fetchFn: (_j = this._options) === null || _j === void 0 ? void 0 : _j.fetch,
                logging: this._options.logging,
            });
            if (_response.ok) {
                return { data: undefined, rawResponse: _response.rawResponse };
            }
            if (_response.error.reason === "status-code") {
                switch (_response.error.statusCode) {
                    case 400:
                        throw new Brevo.BadRequestError(_response.error.body, _response.rawResponse);
                    case 404:
                        throw new Brevo.NotFoundError(_response.error.body, _response.rawResponse);
                    default:
                        throw new errors.BrevoError({
                            statusCode: _response.error.statusCode,
                            body: _response.error.body,
                            rawResponse: _response.rawResponse,
                        });
                }
            }
            return handleNonStatusCodeError(_response.error, _response.rawResponse, "PUT", "/feeds/{uuid}");
        });
    }
    /**
     * Deletes an external feed from your Brevo account.
     *
     * **Use this to:**
     * - Remove external feeds that are no longer needed
     * - Clean up unused data sources
     * - Remove test or outdated feeds
     * - Maintain organized feed library
     *
     * **Key information returned:**
     * - Success confirmation message
     *
     * **Important considerations:**
     * - This action is PERMANENT and cannot be undone
     * - Feed configuration and history will be lost
     * - Check if feed is used in active campaigns before deletion
     * - Remove feed references from email templates
     * - Consider deactivating instead of deleting if unsure
     *
     * @param {Brevo.DeleteExternalFeedRequest} request
     * @param {ExternalFeedsClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link Brevo.BadRequestError}
     * @throws {@link Brevo.NotFoundError}
     *
     * @example
     *     await client.externalFeeds.deleteExternalFeed({
     *         uuid: "b1c2d3e4-f5a6-47b8-89c0-d1e2f3a4b5c6"
     *     })
     */
    deleteExternalFeed(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__deleteExternalFeed(request, requestOptions));
    }
    __deleteExternalFeed(request, requestOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a, _b, _c, _d, _e, _f, _g, _h, _j;
            const { uuid } = request;
            const _authRequest = yield this._options.authProvider.getAuthRequest();
            const _headers = mergeHeaders(_authRequest.headers, (_a = this._options) === null || _a === void 0 ? void 0 : _a.headers, requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.headers);
            const _response = yield core.fetcher({
                url: core.url.join((_c = (_b = (yield core.Supplier.get(this._options.baseUrl))) !== null && _b !== void 0 ? _b : (yield core.Supplier.get(this._options.environment))) !== null && _c !== void 0 ? _c : environments.BrevoEnvironment.Default, `feeds/${core.url.encodePathParam(uuid)}`),
                method: "DELETE",
                headers: _headers,
                queryString: core.url.queryBuilder().mergeAdditional(requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.queryParams).build(),
                timeoutMs: ((_f = (_d = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.timeoutInSeconds) !== null && _d !== void 0 ? _d : (_e = this._options) === null || _e === void 0 ? void 0 : _e.timeoutInSeconds) !== null && _f !== void 0 ? _f : 60) * 1000,
                maxRetries: (_g = requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.maxRetries) !== null && _g !== void 0 ? _g : (_h = this._options) === null || _h === void 0 ? void 0 : _h.maxRetries,
                abortSignal: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.abortSignal,
                fetchFn: (_j = this._options) === null || _j === void 0 ? void 0 : _j.fetch,
                logging: this._options.logging,
            });
            if (_response.ok) {
                return { data: undefined, rawResponse: _response.rawResponse };
            }
            if (_response.error.reason === "status-code") {
                switch (_response.error.statusCode) {
                    case 400:
                        throw new Brevo.BadRequestError(_response.error.body, _response.rawResponse);
                    case 404:
                        throw new Brevo.NotFoundError(_response.error.body, _response.rawResponse);
                    default:
                        throw new errors.BrevoError({
                            statusCode: _response.error.statusCode,
                            body: _response.error.body,
                            rawResponse: _response.rawResponse,
                        });
                }
            }
            return handleNonStatusCodeError(_response.error, _response.rawResponse, "DELETE", "/feeds/{uuid}");
        });
    }
}
