// @getbrevo/brevo/dist/cjs/api/resources/emailCampaigns/types/GetEmailCampaignResponse.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetEmailCampaignResponse = void 0;
var GetEmailCampaignResponse;
(function (GetEmailCampaignResponse) {
    /** Status of the campaign */
    GetEmailCampaignResponse.Status = {
        Draft: "draft",
        Sent: "sent",
        Archive: "archive",
        Queued: "queued",
        Suspended: "suspended",
        InProcess: "in_process",
        InReview: "in_review",
        Cancelling: "cancelling",
        Cancelled: "cancelled",
    };
    /** Type of campaign */
    GetEmailCampaignResponse.Type = {
        Classic: "classic",
        Trigger: "trigger",
    };
    let EmailExpirationDate;
    (function (EmailExpirationDate) {
        /** Unit of the duration */
        EmailExpirationDate.Unit = {
            Days: "days",
            Weeks: "weeks",
            Months: "months",
        };
    })(EmailExpirationDate = GetEmailCampaignResponse.EmailExpirationDate || (GetEmailCampaignResponse.EmailExpirationDate = {}));
})(GetEmailCampaignResponse || (exports.GetEmailCampaignResponse = GetEmailCampaignResponse = {}));
