// @getbrevo/brevo/dist/cjs/api/resources/emailCampaigns/types/GetEmailCampaignsResponse.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetEmailCampaignsResponse = void 0;
var GetEmailCampaignsResponse;
(function (GetEmailCampaignsResponse) {
    let Campaigns;
    (function (Campaigns) {
        let Item;
        (function (Item) {
            /** Status of the campaign */
            Item.Status = {
                Draft: "draft",
                Sent: "sent",
                Archive: "archive",
                Queued: "queued",
                Suspended: "suspended",
                InProcess: "in_process",
                InReview: "in_review",
                Cancelling: "cancelling",
                Cancelled: "cancelled",
            };
            /** Type of campaign */
            Item.Type = {
                Classic: "classic",
                Trigger: "trigger",
            };
            let EmailExpirationDate;
            (function (EmailExpirationDate) {
                /** Unit of the duration */
                EmailExpirationDate.Unit = {
                    Days: "days",
                    Weeks: "weeks",
                    Months: "months",
                };
            })(EmailExpirationDate = Item.EmailExpirationDate || (Item.EmailExpirationDate = {}));
        })(Item = Campaigns.Item || (Campaigns.Item = {}));
    })(Campaigns = GetEmailCampaignsResponse.Campaigns || (GetEmailCampaignsResponse.Campaigns = {}));
})(GetEmailCampaignsResponse || (exports.GetEmailCampaignsResponse = GetEmailCampaignsResponse = {}));
