// @getbrevo/brevo/dist/cjs/api/types/GetExtendedCampaignOverview.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetExtendedCampaignOverview = void 0;
var GetExtendedCampaignOverview;
(function (GetExtendedCampaignOverview) {
    /** Status of the campaign */
    GetExtendedCampaignOverview.Status = {
        Draft: "draft",
        Sent: "sent",
        Archive: "archive",
        Queued: "queued",
        Suspended: "suspended",
        InProcess: "in_process",
        InReview: "in_review",
        Cancelling: "cancelling",
        Cancelled: "cancelled",
    };
    /** Type of campaign */
    GetExtendedCampaignOverview.Type = {
        Classic: "classic",
        Trigger: "trigger",
    };
    let EmailExpirationDate;
    (function (EmailExpirationDate) {
        /** Unit of the duration */
        EmailExpirationDate.Unit = {
            Days: "days",
            Weeks: "weeks",
            Months: "months",
        };
    })(EmailExpirationDate = GetExtendedCampaignOverview.EmailExpirationDate || (GetExtendedCampaignOverview.EmailExpirationDate = {}));
})(GetExtendedCampaignOverview || (exports.GetExtendedCampaignOverview = GetExtendedCampaignOverview = {}));
