// @getbrevo/brevo/dist/cjs/api/types/Inviteuser.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Inviteuser = void 0;
var Inviteuser;
(function (Inviteuser) {
    let Privileges;
    (function (Privileges) {
        let Item;
        (function (Item_1) {
            /** Feature name */
            Item_1.Feature = {
                EmailCampaigns: "email_campaigns",
                SmsCampaigns: "sms_campaigns",
                Contacts: "contacts",
                Templates: "templates",
                Workflows: "workflows",
                LandingPages: "landing_pages",
                TransactionalEmails: "transactional_emails",
                SmtpApi: "smtp_api",
                UserManagement: "user_management",
                SalesPlatform: "sales_platform",
                Phone: "phone",
                Conversations: "conversations",
                SendersDomainsDedicatedIps: "senders_domains_dedicated_ips",
                PushNotifications: "push_notifications",
                Companies: "companies",
            };
            let Permissions;
            (function (Permissions) {
                Permissions.Item = {
                    CreateEditDelete: "create_edit_delete",
                    SendScheduleSuspend: "send_schedule_suspend",
                    View: "view",
                    Import: "import",
                    Export: "export",
                    ListAndAttributes: "list_and_attributes",
                    Forms: "forms",
                    ActivateDeactivate: "activate_deactivate",
                    ActivateDeactivatePause: "activate_deactivate_pause",
                    Settings: "settings",
                    SchedulePause: "schedule_pause",
                    All: "all",
                    Logs: "logs",
                    Access: "access",
                    Assign: "assign",
                    Configure: "configure",
                    CreateEditDeals: "create_edit_deals",
                    DeleteDeals: "delete_deals",
                    ManageOthersDealsTasks: "manage_others_deals_tasks",
                    ManageOwnedCompanies: "manage_owned_companies",
                    ManageOthersCompanies: "manage_others_companies",
                    Reports: "reports",
                    SendersManagement: "senders_management",
                    DomainsManagement: "domains_management",
                    DedicatedIpsManagement: "dedicated_ips_management",
                    Send: "send",
                    Smtp: "smtp",
                    ApiKeys: "api_keys",
                    AuthorizedIps: "authorized_ips",
                    None: "none",
                };
            })(Permissions = Item_1.Permissions || (Item_1.Permissions = {}));
        })(Item = Privileges.Item || (Privileges.Item = {}));
    })(Privileges = Inviteuser.Privileges || (Inviteuser.Privileges = {}));
})(Inviteuser || (exports.Inviteuser = Inviteuser = {}));
