// @getbrevo/brevo/dist/esm/Client.mjs
// This file was auto-generated by Fern from our API Definition.
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { AccountClient } from "./api/resources/account/client/Client.mjs";
import { BalanceClient } from "./api/resources/balance/client/Client.mjs";
import { CompaniesClient } from "./api/resources/companies/client/Client.mjs";
import { ConsentGroupsClient } from "./api/resources/consentGroups/client/Client.mjs";
import { ContactsClient } from "./api/resources/contacts/client/Client.mjs";
import { ConversationsClient } from "./api/resources/conversations/client/Client.mjs";
import { CouponsClient } from "./api/resources/coupons/client/Client.mjs";
import { CustomObjectsClient } from "./api/resources/customObjects/client/Client.mjs";
import { DealsClient } from "./api/resources/deals/client/Client.mjs";
import { DomainsClient } from "./api/resources/domains/client/Client.mjs";
import { EcommerceClient } from "./api/resources/ecommerce/client/Client.mjs";
import { EmailCampaignsClient } from "./api/resources/emailCampaigns/client/Client.mjs";
import { EventClient } from "./api/resources/event/client/Client.mjs";
import { ExternalFeedsClient } from "./api/resources/externalFeeds/client/Client.mjs";
import { FilesClient } from "./api/resources/files/client/Client.mjs";
import { InboundParsingClient } from "./api/resources/inboundParsing/client/Client.mjs";
import { MasterAccountClient } from "./api/resources/masterAccount/client/Client.mjs";
import { NotesClient } from "./api/resources/notes/client/Client.mjs";
import { PaymentsClient } from "./api/resources/payments/client/Client.mjs";
import { ProcessClient } from "./api/resources/process/client/Client.mjs";
import { ProgramClient } from "./api/resources/program/client/Client.mjs";
import { RewardClient } from "./api/resources/reward/client/Client.mjs";
import { SendersClient } from "./api/resources/senders/client/Client.mjs";
import { SmsCampaignsClient } from "./api/resources/smsCampaigns/client/Client.mjs";
import { SmsTemplatesClient } from "./api/resources/smsTemplates/client/Client.mjs";
import { TasksClient } from "./api/resources/tasks/client/Client.mjs";
import { TierClient } from "./api/resources/tier/client/Client.mjs";
import { TransactionalEmailsClient } from "./api/resources/transactionalEmails/client/Client.mjs";
import { TransactionalSmsClient } from "./api/resources/transactionalSms/client/Client.mjs";
import { TransactionalWhatsAppClient } from "./api/resources/transactionalWhatsApp/client/Client.mjs";
import { UserClient } from "./api/resources/user/client/Client.mjs";
import { WalletClient } from "./api/resources/wallet/client/Client.mjs";
import { WebhooksClient } from "./api/resources/webhooks/client/Client.mjs";
import { WhatsAppCampaignsClient } from "./api/resources/whatsAppCampaigns/client/Client.mjs";
import { normalizeClientOptionsWithAuth } from "./BaseClient.mjs";
import * as core from "./core/index.mjs";
export class BrevoClient {
    constructor(options) {
        this._options = normalizeClientOptionsWithAuth(options);
    }
    get account() {
        var _a;
        return ((_a = this._account) !== null && _a !== void 0 ? _a : (this._account = new AccountClient(this._options)));
    }
    get masterAccount() {
        var _a;
        return ((_a = this._masterAccount) !== null && _a !== void 0 ? _a : (this._masterAccount = new MasterAccountClient(this._options)));
    }
    get user() {
        var _a;
        return ((_a = this._user) !== null && _a !== void 0 ? _a : (this._user = new UserClient(this._options)));
    }
    get process() {
        var _a;
        return ((_a = this._process) !== null && _a !== void 0 ? _a : (this._process = new ProcessClient(this._options)));
    }
    get senders() {
        var _a;
        return ((_a = this._senders) !== null && _a !== void 0 ? _a : (this._senders = new SendersClient(this._options)));
    }
    get domains() {
        var _a;
        return ((_a = this._domains) !== null && _a !== void 0 ? _a : (this._domains = new DomainsClient(this._options)));
    }
    get webhooks() {
        var _a;
        return ((_a = this._webhooks) !== null && _a !== void 0 ? _a : (this._webhooks = new WebhooksClient(this._options)));
    }
    get externalFeeds() {
        var _a;
        return ((_a = this._externalFeeds) !== null && _a !== void 0 ? _a : (this._externalFeeds = new ExternalFeedsClient(this._options)));
    }
    get customObjects() {
        var _a;
        return ((_a = this._customObjects) !== null && _a !== void 0 ? _a : (this._customObjects = new CustomObjectsClient(this._options)));
    }
    get contacts() {
        var _a;
        return ((_a = this._contacts) !== null && _a !== void 0 ? _a : (this._contacts = new ContactsClient(this._options)));
    }
    get consentGroups() {
        var _a;
        return ((_a = this._consentGroups) !== null && _a !== void 0 ? _a : (this._consentGroups = new ConsentGroupsClient(this._options)));
    }
    get conversations() {
        var _a;
        return ((_a = this._conversations) !== null && _a !== void 0 ? _a : (this._conversations = new ConversationsClient(this._options)));
    }
    get ecommerce() {
        var _a;
        return ((_a = this._ecommerce) !== null && _a !== void 0 ? _a : (this._ecommerce = new EcommerceClient(this._options)));
    }
    get coupons() {
        var _a;
        return ((_a = this._coupons) !== null && _a !== void 0 ? _a : (this._coupons = new CouponsClient(this._options)));
    }
    get payments() {
        var _a;
        return ((_a = this._payments) !== null && _a !== void 0 ? _a : (this._payments = new PaymentsClient(this._options)));
    }
    get event() {
        var _a;
        return ((_a = this._event) !== null && _a !== void 0 ? _a : (this._event = new EventClient(this._options)));
    }
    get inboundParsing() {
        var _a;
        return ((_a = this._inboundParsing) !== null && _a !== void 0 ? _a : (this._inboundParsing = new InboundParsingClient(this._options)));
    }
    get balance() {
        var _a;
        return ((_a = this._balance) !== null && _a !== void 0 ? _a : (this._balance = new BalanceClient(this._options)));
    }
    get program() {
        var _a;
        return ((_a = this._program) !== null && _a !== void 0 ? _a : (this._program = new ProgramClient(this._options)));
    }
    get reward() {
        var _a;
        return ((_a = this._reward) !== null && _a !== void 0 ? _a : (this._reward = new RewardClient(this._options)));
    }
    get tier() {
        var _a;
        return ((_a = this._tier) !== null && _a !== void 0 ? _a : (this._tier = new TierClient(this._options)));
    }
    get wallet() {
        var _a;
        return ((_a = this._wallet) !== null && _a !== void 0 ? _a : (this._wallet = new WalletClient(this._options)));
    }
    get emailCampaigns() {
        var _a;
        return ((_a = this._emailCampaigns) !== null && _a !== void 0 ? _a : (this._emailCampaigns = new EmailCampaignsClient(this._options)));
    }
    get smsCampaigns() {
        var _a;
        return ((_a = this._smsCampaigns) !== null && _a !== void 0 ? _a : (this._smsCampaigns = new SmsCampaignsClient(this._options)));
    }
    get whatsAppCampaigns() {
        var _a;
        return ((_a = this._whatsAppCampaigns) !== null && _a !== void 0 ? _a : (this._whatsAppCampaigns = new WhatsAppCampaignsClient(this._options)));
    }
    get companies() {
        var _a;
        return ((_a = this._companies) !== null && _a !== void 0 ? _a : (this._companies = new CompaniesClient(this._options)));
    }
    get deals() {
        var _a;
        return ((_a = this._deals) !== null && _a !== void 0 ? _a : (this._deals = new DealsClient(this._options)));
    }
    get files() {
        var _a;
        return ((_a = this._files) !== null && _a !== void 0 ? _a : (this._files = new FilesClient(this._options)));
    }
    get notes() {
        var _a;
        return ((_a = this._notes) !== null && _a !== void 0 ? _a : (this._notes = new NotesClient(this._options)));
    }
    get tasks() {
        var _a;
        return ((_a = this._tasks) !== null && _a !== void 0 ? _a : (this._tasks = new TasksClient(this._options)));
    }
    get transactionalWhatsApp() {
        var _a;
        return ((_a = this._transactionalWhatsApp) !== null && _a !== void 0 ? _a : (this._transactionalWhatsApp = new TransactionalWhatsAppClient(this._options)));
    }
    get transactionalEmails() {
        var _a;
        return ((_a = this._transactionalEmails) !== null && _a !== void 0 ? _a : (this._transactionalEmails = new TransactionalEmailsClient(this._options)));
    }
    get transactionalSms() {
        var _a;
        return ((_a = this._transactionalSms) !== null && _a !== void 0 ? _a : (this._transactionalSms = new TransactionalSmsClient(this._options)));
    }
    get smsTemplates() {
        var _a;
        return ((_a = this._smsTemplates) !== null && _a !== void 0 ? _a : (this._smsTemplates = new SmsTemplatesClient(this._options)));
    }
    /**
     * Make a passthrough request using the SDK's configured auth, retry, logging, etc.
     * This is useful for making requests to endpoints not yet supported in the SDK.
     * The input can be a URL string, URL object, or Request object. Relative paths are resolved against the configured base URL.
     *
     * @param {Request | string | URL} input - The URL, path, or Request object.
     * @param {RequestInit} init - Standard fetch RequestInit options.
     * @param {core.PassthroughRequest.RequestOptions} requestOptions - Per-request overrides (timeout, retries, headers, abort signal).
     * @returns {Promise<Response>} A standard Response object.
     */
    fetch(input, init, requestOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a;
            return core.makePassthroughRequest(input, init, {
                baseUrl: (_a = this._options.baseUrl) !== null && _a !== void 0 ? _a : this._options.environment,
                headers: this._options.headers,
                timeoutInSeconds: this._options.timeoutInSeconds,
                maxRetries: this._options.maxRetries,
                fetch: this._options.fetch,
                logging: this._options.logging,
                getAuthHeaders: () => __awaiter(this, void 0, void 0, function* () { return (yield this._options.authProvider.getAuthRequest()).headers; }),
            }, requestOptions);
        });
    }
}
