// @getbrevo/brevo/dist/cjs/api/resources/whatsAppCampaigns/types/GetWhatsAppCampaignsResponse.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetWhatsAppCampaignsResponse = void 0;
var GetWhatsAppCampaignsResponse;
(function (GetWhatsAppCampaignsResponse) {
    let Campaigns;
    (function (Campaigns) {
        let Item;
        (function (Item) {
            /** Status of the Whatsapp Campaign */
            Item.CampaignStatus = {
                Draft: "draft",
                Scheduled: "scheduled",
                Pending: "pending",
                Approved: "approved",
                Running: "running",
                Suspended: "suspended",
                Rejected: "rejected",
                Sent: "sent",
            };
        })(Item = Campaigns.Item || (Campaigns.Item = {}));
    })(Campaigns = GetWhatsAppCampaignsResponse.Campaigns || (GetWhatsAppCampaignsResponse.Campaigns = {}));
})(GetWhatsAppCampaignsResponse || (exports.GetWhatsAppCampaignsResponse = GetWhatsAppCampaignsResponse = {}));
