// @getbrevo/brevo/dist/cjs/api/resources/whatsAppCampaigns/types/GetWhatsAppConfigResponse.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetWhatsAppConfigResponse = void 0;
var GetWhatsAppConfigResponse;
(function (GetWhatsAppConfigResponse) {
    /** Status of the name associated with WhatsApp Phone number */
    GetWhatsAppConfigResponse.PhoneNumberNameStatus = {
        Approved: "APPROVED",
        Pending: "PENDING",
        Rejected: "REJECTED",
    };
    /** Quality status of phone number associated with WhatsApp account. There are three quality ratings. example - **High (GREEN) , Medium (YELLOW) and Low(RED)** */
    GetWhatsAppConfigResponse.PhoneNumberQuality = {
        Green: "GREEN",
        Yellow: "YELLOW",
        Red: "RED",
    };
    /** Status information related to WhatsApp Api account */
    GetWhatsAppConfigResponse.WhatsappBusinessAccountStatus = {
        Approved: "APPROVED",
        Pending: "PENDING",
        Rejected: "REJECTED",
    };
})(GetWhatsAppConfigResponse || (exports.GetWhatsAppConfigResponse = GetWhatsAppConfigResponse = {}));
