// @getbrevo/brevo/dist/esm/auth/HeaderAuthProvider.mjs
// This file was auto-generated by Fern from our API Definition.
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import * as core from "../core/index.mjs";
import * as errors from "../errors/index.mjs";
const PARAM_KEY = "apiKey";
const HEADER_NAME = "api-key";
export class HeaderAuthProvider {
    constructor(options) {
        this.options = options;
    }
    static canCreate(options) {
        return (options === null || options === void 0 ? void 0 : options[PARAM_KEY]) != null;
    }
    getAuthRequest() {
        return __awaiter(this, arguments, void 0, function* ({ endpointMetadata, } = {}) {
            const headerValue = yield core.Supplier.get(this.options[PARAM_KEY]);
            if (headerValue == null) {
                throw new errors.BrevoError({
                    message: HeaderAuthProvider.AUTH_CONFIG_ERROR_MESSAGE,
                });
            }
            return {
                headers: { [HEADER_NAME]: headerValue },
            };
        });
    }
}
(function (HeaderAuthProvider) {
    HeaderAuthProvider.AUTH_SCHEME = "api-key";
    HeaderAuthProvider.AUTH_CONFIG_ERROR_MESSAGE = `Please provide '${PARAM_KEY}' when initializing the client`;
    function createInstance(options) {
        return new HeaderAuthProvider(options);
    }
    HeaderAuthProvider.createInstance = createInstance;
})(HeaderAuthProvider || (HeaderAuthProvider = {}));
