// @getbrevo/brevo/dist/esm/api/resources/contacts/types/GetAttributesResponse.mjs
// This file was auto-generated by Fern from our API Definition.
export var GetAttributesResponse;
(function (GetAttributesResponse) {
    let Attributes;
    (function (Attributes) {
        let Item;
        (function (Item) {
            /** Category of the attribute */
            Item.Category = {
                Normal: "normal",
                Transactional: "transactional",
                Category: "category",
                Calculated: "calculated",
                Global: "global",
            };
            /** Type of the attribute */
            Item.Type = {
                Text: "text",
                Date: "date",
                Float: "float",
                Id: "id",
                Boolean: "boolean",
                MultipleChoice: "multiple-choice",
                User: "user",
            };
        })(Item = Attributes.Item || (Attributes.Item = {}));
    })(Attributes = GetAttributesResponse.Attributes || (GetAttributesResponse.Attributes = {}));
})(GetAttributesResponse || (GetAttributesResponse = {}));
