// @getbrevo/brevo/dist/cjs/api/resources/webhooks/client/requests/CreateWebhookRequest.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateWebhookRequest = void 0;
var CreateWebhookRequest;
(function (CreateWebhookRequest) {
    /** Channel of the webhook */
    CreateWebhookRequest.Channel = {
        Email: "email",
        Sms: "sms",
    };
    let Events;
    (function (Events) {
        Events.Item = {
            Sent: "sent",
            HardBounce: "hardBounce",
            SoftBounce: "softBounce",
            Blocked: "blocked",
            Spam: "spam",
            Delivered: "delivered",
            Request: "request",
            Click: "click",
            Invalid: "invalid",
            Deferred: "deferred",
            Opened: "opened",
            UniqueOpened: "uniqueOpened",
            Unsubscribed: "unsubscribed",
            ListAddition: "listAddition",
            ContactUpdated: "contactUpdated",
            ContactDeleted: "contactDeleted",
            InboundEmailProcessed: "inboundEmailProcessed",
            Reply: "reply",
        };
    })(Events = CreateWebhookRequest.Events || (CreateWebhookRequest.Events = {}));
    /** Type of the webhook */
    CreateWebhookRequest.Type = {
        Transactional: "transactional",
        Marketing: "marketing",
        Inbound: "inbound",
    };
})(CreateWebhookRequest || (exports.CreateWebhookRequest = CreateWebhookRequest = {}));
