// @getbrevo/brevo/dist/cjs/Client.js
"use strict";
// This file was auto-generated by Fern from our API Definition.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BrevoClient = void 0;
const Client_js_1 = require("./api/resources/account/client/Client.js");
const Client_js_2 = require("./api/resources/balance/client/Client.js");
const Client_js_3 = require("./api/resources/companies/client/Client.js");
const Client_js_4 = require("./api/resources/consentGroups/client/Client.js");
const Client_js_5 = require("./api/resources/contacts/client/Client.js");
const Client_js_6 = require("./api/resources/conversations/client/Client.js");
const Client_js_7 = require("./api/resources/coupons/client/Client.js");
const Client_js_8 = require("./api/resources/customObjects/client/Client.js");
const Client_js_9 = require("./api/resources/deals/client/Client.js");
const Client_js_10 = require("./api/resources/domains/client/Client.js");
const Client_js_11 = require("./api/resources/ecommerce/client/Client.js");
const Client_js_12 = require("./api/resources/emailCampaigns/client/Client.js");
const Client_js_13 = require("./api/resources/event/client/Client.js");
const Client_js_14 = require("./api/resources/externalFeeds/client/Client.js");
const Client_js_15 = require("./api/resources/files/client/Client.js");
const Client_js_16 = require("./api/resources/inboundParsing/client/Client.js");
const Client_js_17 = require("./api/resources/masterAccount/client/Client.js");
const Client_js_18 = require("./api/resources/notes/client/Client.js");
const Client_js_19 = require("./api/resources/payments/client/Client.js");
const Client_js_20 = require("./api/resources/process/client/Client.js");
const Client_js_21 = require("./api/resources/program/client/Client.js");
const Client_js_22 = require("./api/resources/reward/client/Client.js");
const Client_js_23 = require("./api/resources/senders/client/Client.js");
const Client_js_24 = require("./api/resources/smsCampaigns/client/Client.js");
const Client_js_25 = require("./api/resources/smsTemplates/client/Client.js");
const Client_js_26 = require("./api/resources/tasks/client/Client.js");
const Client_js_27 = require("./api/resources/tier/client/Client.js");
const Client_js_28 = require("./api/resources/transactionalEmails/client/Client.js");
const Client_js_29 = require("./api/resources/transactionalSms/client/Client.js");
const Client_js_30 = require("./api/resources/transactionalWhatsApp/client/Client.js");
const Client_js_31 = require("./api/resources/user/client/Client.js");
const Client_js_32 = require("./api/resources/wallet/client/Client.js");
const Client_js_33 = require("./api/resources/webhooks/client/Client.js");
const Client_js_34 = require("./api/resources/whatsAppCampaigns/client/Client.js");
const BaseClient_js_1 = require("./BaseClient.js");
const core = __importStar(require("./core/index.js"));
class BrevoClient {
    constructor(options) {
        this._options = (0, BaseClient_js_1.normalizeClientOptionsWithAuth)(options);
    }
    get account() {
        var _a;
        return ((_a = this._account) !== null && _a !== void 0 ? _a : (this._account = new Client_js_1.AccountClient(this._options)));
    }
    get masterAccount() {
        var _a;
        return ((_a = this._masterAccount) !== null && _a !== void 0 ? _a : (this._masterAccount = new Client_js_17.MasterAccountClient(this._options)));
    }
    get user() {
        var _a;
        return ((_a = this._user) !== null && _a !== void 0 ? _a : (this._user = new Client_js_31.UserClient(this._options)));
    }
    get process() {
        var _a;
        return ((_a = this._process) !== null && _a !== void 0 ? _a : (this._process = new Client_js_20.ProcessClient(this._options)));
    }
    get senders() {
        var _a;
        return ((_a = this._senders) !== null && _a !== void 0 ? _a : (this._senders = new Client_js_23.SendersClient(this._options)));
    }
    get domains() {
        var _a;
        return ((_a = this._domains) !== null && _a !== void 0 ? _a : (this._domains = new Client_js_10.DomainsClient(this._options)));
    }
    get webhooks() {
        var _a;
        return ((_a = this._webhooks) !== null && _a !== void 0 ? _a : (this._webhooks = new Client_js_33.WebhooksClient(this._options)));
    }
    get externalFeeds() {
        var _a;
        return ((_a = this._externalFeeds) !== null && _a !== void 0 ? _a : (this._externalFeeds = new Client_js_14.ExternalFeedsClient(this._options)));
    }
    get customObjects() {
        var _a;
        return ((_a = this._customObjects) !== null && _a !== void 0 ? _a : (this._customObjects = new Client_js_8.CustomObjectsClient(this._options)));
    }
    get contacts() {
        var _a;
        return ((_a = this._contacts) !== null && _a !== void 0 ? _a : (this._contacts = new Client_js_5.ContactsClient(this._options)));
    }
    get consentGroups() {
        var _a;
        return ((_a = this._consentGroups) !== null && _a !== void 0 ? _a : (this._consentGroups = new Client_js_4.ConsentGroupsClient(this._options)));
    }
    get conversations() {
        var _a;
        return ((_a = this._conversations) !== null && _a !== void 0 ? _a : (this._conversations = new Client_js_6.ConversationsClient(this._options)));
    }
    get ecommerce() {
        var _a;
        return ((_a = this._ecommerce) !== null && _a !== void 0 ? _a : (this._ecommerce = new Client_js_11.EcommerceClient(this._options)));
    }
    get coupons() {
        var _a;
        return ((_a = this._coupons) !== null && _a !== void 0 ? _a : (this._coupons = new Client_js_7.CouponsClient(this._options)));
    }
    get payments() {
        var _a;
        return ((_a = this._payments) !== null && _a !== void 0 ? _a : (this._payments = new Client_js_19.PaymentsClient(this._options)));
    }
    get event() {
        var _a;
        return ((_a = this._event) !== null && _a !== void 0 ? _a : (this._event = new Client_js_13.EventClient(this._options)));
    }
    get inboundParsing() {
        var _a;
        return ((_a = this._inboundParsing) !== null && _a !== void 0 ? _a : (this._inboundParsing = new Client_js_16.InboundParsingClient(this._options)));
    }
    get balance() {
        var _a;
        return ((_a = this._balance) !== null && _a !== void 0 ? _a : (this._balance = new Client_js_2.BalanceClient(this._options)));
    }
    get program() {
        var _a;
        return ((_a = this._program) !== null && _a !== void 0 ? _a : (this._program = new Client_js_21.ProgramClient(this._options)));
    }
    get reward() {
        var _a;
        return ((_a = this._reward) !== null && _a !== void 0 ? _a : (this._reward = new Client_js_22.RewardClient(this._options)));
    }
    get tier() {
        var _a;
        return ((_a = this._tier) !== null && _a !== void 0 ? _a : (this._tier = new Client_js_27.TierClient(this._options)));
    }
    get wallet() {
        var _a;
        return ((_a = this._wallet) !== null && _a !== void 0 ? _a : (this._wallet = new Client_js_32.WalletClient(this._options)));
    }
    get emailCampaigns() {
        var _a;
        return ((_a = this._emailCampaigns) !== null && _a !== void 0 ? _a : (this._emailCampaigns = new Client_js_12.EmailCampaignsClient(this._options)));
    }
    get smsCampaigns() {
        var _a;
        return ((_a = this._smsCampaigns) !== null && _a !== void 0 ? _a : (this._smsCampaigns = new Client_js_24.SmsCampaignsClient(this._options)));
    }
    get whatsAppCampaigns() {
        var _a;
        return ((_a = this._whatsAppCampaigns) !== null && _a !== void 0 ? _a : (this._whatsAppCampaigns = new Client_js_34.WhatsAppCampaignsClient(this._options)));
    }
    get companies() {
        var _a;
        return ((_a = this._companies) !== null && _a !== void 0 ? _a : (this._companies = new Client_js_3.CompaniesClient(this._options)));
    }
    get deals() {
        var _a;
        return ((_a = this._deals) !== null && _a !== void 0 ? _a : (this._deals = new Client_js_9.DealsClient(this._options)));
    }
    get files() {
        var _a;
        return ((_a = this._files) !== null && _a !== void 0 ? _a : (this._files = new Client_js_15.FilesClient(this._options)));
    }
    get notes() {
        var _a;
        return ((_a = this._notes) !== null && _a !== void 0 ? _a : (this._notes = new Client_js_18.NotesClient(this._options)));
    }
    get tasks() {
        var _a;
        return ((_a = this._tasks) !== null && _a !== void 0 ? _a : (this._tasks = new Client_js_26.TasksClient(this._options)));
    }
    get transactionalWhatsApp() {
        var _a;
        return ((_a = this._transactionalWhatsApp) !== null && _a !== void 0 ? _a : (this._transactionalWhatsApp = new Client_js_30.TransactionalWhatsAppClient(this._options)));
    }
    get transactionalEmails() {
        var _a;
        return ((_a = this._transactionalEmails) !== null && _a !== void 0 ? _a : (this._transactionalEmails = new Client_js_28.TransactionalEmailsClient(this._options)));
    }
    get transactionalSms() {
        var _a;
        return ((_a = this._transactionalSms) !== null && _a !== void 0 ? _a : (this._transactionalSms = new Client_js_29.TransactionalSmsClient(this._options)));
    }
    get smsTemplates() {
        var _a;
        return ((_a = this._smsTemplates) !== null && _a !== void 0 ? _a : (this._smsTemplates = new Client_js_25.SmsTemplatesClient(this._options)));
    }
    /**
     * Make a passthrough request using the SDK's configured auth, retry, logging, etc.
     * This is useful for making requests to endpoints not yet supported in the SDK.
     * The input can be a URL string, URL object, or Request object. Relative paths are resolved against the configured base URL.
     *
     * @param {Request | string | URL} input - The URL, path, or Request object.
     * @param {RequestInit} init - Standard fetch RequestInit options.
     * @param {core.PassthroughRequest.RequestOptions} requestOptions - Per-request overrides (timeout, retries, headers, abort signal).
     * @returns {Promise<Response>} A standard Response object.
     */
    fetch(input, init, requestOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a;
            return core.makePassthroughRequest(input, init, {
                baseUrl: (_a = this._options.baseUrl) !== null && _a !== void 0 ? _a : this._options.environment,
                headers: this._options.headers,
                timeoutInSeconds: this._options.timeoutInSeconds,
                maxRetries: this._options.maxRetries,
                fetch: this._options.fetch,
                logging: this._options.logging,
                getAuthHeaders: () => __awaiter(this, void 0, void 0, function* () { return (yield this._options.authProvider.getAuthRequest()).headers; }),
            }, requestOptions);
        });
    }
}
exports.BrevoClient = BrevoClient;
